{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find stats\n",
    "\n",
    "for hichip figure plus supplements\n",
    "\n",
    "05/08/2020\n",
    "\n",
    "05/12/2020 - update how we find the promoter regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = '../data/processed/fig2_hichip'\n",
    "if not os.path.exists(save_dir):\n",
    "    os.makedirs(save_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "loop_files = glob.glob('../data/interim/merged/loops/*.loops.csv')\n",
    "anchor_annon_dir = '../data/interim/annon/anchor_atac'\n",
    "anchor_annon_files = glob.glob('../data/interim/annon/anchor_atac/*_annon.bed')\n",
    "promoter_annon_dir = '../data/interim/annon/promoter_anchors/'\n",
    "promoter_annon_files = glob.glob('../data/interim/annon/promoter_anchors/promoter_*_annon.bed')\n",
    "# anchor_annon_files\n",
    "anchor_dir = '../data/interim/merged/anchors/'\n",
    "anchor_files = glob.glob('../data/interim/merged/anchors/*.anchors.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "normal_tissues = [ \"Astrocytes\",\"H9_D0\",\"H9_D10\",\"H9_D2\",\"H9_D28\",\"H9_D4\",\"SL_D0\",\"SL_D2\",\"SLC_D0\",\"SLC_D2\"]\n",
    "normal_tissues = [ \"Astrocytes\",\"H9_D0\",\"H9_D10\",\"H9_D2\",\"H9_D28\",\"SL_D0\",\"SL_D2\",\"SLC_D0\",\"SLC_D2\"]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Find Stats of of distribution of loop width"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Astrocytes (1839205, 2)\n",
      "H9_D0 (3157921, 2)\n",
      "H9_D10 (3870454, 2)\n",
      "H9_D2 (10590560, 2)\n",
      "H9_D28 (11847519, 2)\n",
      "SLC_D0 (14506601, 2)\n",
      "SLC_D2 (18788756, 2)\n",
      "SL_D0 (19976355, 2)\n",
      "SL_D2 (25300784, 2)\n",
      "CPU times: user 3min 54s, sys: 5.95 s, total: 4min\n",
      "Wall time: 3min 9s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "loop_width_df = pd.DataFrame(columns=['loop_width','tissue'])\n",
    "for loop_file in sorted(loop_files):\n",
    "    loop_filename = os.path.basename(loop_file)\n",
    "    tissue = loop_filename.split('.')[0]\n",
    "    loop_df = pd.read_csv(loop_file, index_col=0)\n",
    "    loop_df[['chrA','startA','stopA']] = loop_df.source.str.split('_',expand=True)\n",
    "    loop_df[['chrB','startB','stopB']] = loop_df.target.str.split('_',expand=True)\n",
    "    loop_df['loop_width'] = loop_df.startB.map(int) - loop_df.startA.map(int)\n",
    "    loop_df['tissue'] = tissue\n",
    "    loop_width_df = pd.concat([loop_width_df, loop_df[['loop_width', 'tissue']]])\n",
    "    print(tissue, loop_width_df.shape)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(25300784, 2) (25300784, 2)\n"
     ]
    }
   ],
   "source": [
    "loop_width_normal_df = loop_width_df[loop_width_df.tissue.isin(normal_tissues)]\n",
    "print(loop_width_df.shape, loop_width_normal_df.shape)\n",
    "loop_width_df.to_csv(os.path.join(save_dir, 'loop_width_df.csv'))\n",
    "loop_width_normal_df.to_csv(os.path.join(save_dir, 'loop_width_normal_df.csv'))\n",
    "# loop_width_df[:5]\n",
    "# loop_width_df_test = loop_width_df[(loop_width_df.tissue=='Air')|(loop_width_df.tissue=='MC') ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Loop and Anchor statistics\n",
    "\n",
    "for each tissue (split normal and all) get following stats\n",
    "- num unique loops\n",
    "- num unique anchors\n",
    "- num common loops\n",
    "- num common anchors \n",
    "- num loops\n",
    "- num anchors \t\n",
    "\n",
    "summative stats (across all tissues)\n",
    "    - total number of distinct anchors\n",
    "    - total number of distinct loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Astrocytes\n",
      "H9_D0\n",
      "H9_D10\n",
      "H9_D2\n",
      "H9_D28\n",
      "SLC_D0\n",
      "SLC_D2\n",
      "SL_D0\n",
      "SL_D2\n",
      "CPU times: user 12min 56s, sys: 3.82 s, total: 13min\n",
      "Wall time: 12min 16s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# first get set of all loops\n",
    "# all_loops, all_loops_normal = set(),set()\n",
    "# all_anchors, all_anchors_normal = set(),set()\n",
    "\n",
    "tissues = []\n",
    "loops_all_df = pd.DataFrame(columns=['loop_name','tissue'])\n",
    "anchors_all_df = pd.DataFrame(columns=['anchors','tissue'])\n",
    "\n",
    "# tissue_info_dict = {}\n",
    "for loop_file in sorted(loop_files):\n",
    "    loop_filename = os.path.basename(loop_file)\n",
    "    tissue = loop_filename.split('.')[0]\n",
    "    tissues.append(tissue)\n",
    "    anchor_file = os.path.join(anchor_dir, tissue+'.anchors.csv')\n",
    "    loop_df = pd.read_csv(loop_file, index_col=0)\n",
    "    loop_df['loop_name'] = loop_df.apply(lambda row: '::'.join(sorted([row.source,row.target])),axis=1)\n",
    "    loop_df['tissue'] = tissue\n",
    "    \n",
    "    anchor_df = pd.read_csv(anchor_file, index_col=0)\n",
    "    anchor_df['tissue'] = tissue\n",
    "   \n",
    "    loops_all_df = pd.concat([loops_all_df, loop_df[['loop_name', 'tissue']]])\n",
    "    anchors_all_df = pd.concat([anchors_all_df, anchor_df[['anchors', 'tissue']]])\n",
    "    print(tissue)\n",
    "    \n",
    "#     tissue_info_dict[tissue] = [loop_df,anchor_df,anchor_to_count_dict]\n",
    "    \n",
    "#     all_loops  = all_loops.union(set(loop_df.loop_name))\n",
    "#     all_anchors  = all_anchors.union(set(anchor_df.anchors))\n",
    "                                         \n",
    "#     if tissue in normal_tissues:\n",
    "#         all_loops_normal = all_loops_normal.union(set(loop_df.loop_name))\n",
    "#         all_anchors_normal = all_anchors_normal.union(set(anchor_df.anchors))                            \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_loops = loops_all_df.loop_name.unique()\n",
    "all_loops_normal = loops_all_df.loop_name[loops_all_df.tissue.isin(normal_tissues)].unique()\n",
    "all_anchors = anchors_all_df.anchors.unique()\n",
    "all_anchors_normal = anchors_all_df.anchors[anchors_all_df.tissue.isin(normal_tissues)].unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>loop_name</th>\n",
       "      <th>tissue</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr10_100000000_100005000::chr10_100015000_100...</td>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>chr10_100000000_100005000::chr10_100160000_100...</td>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>chr10_100000000_100005000::chr10_100195000_100...</td>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>143</td>\n",
       "      <td>chr10_10000000_10005000::chr10_11035000_11040000</td>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>199</td>\n",
       "      <td>chr10_1000000_1005000::chr10_1020000_1025000</td>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>80042894</td>\n",
       "      <td>chrX_100460000_100465000::chrX_99995000_100000000</td>\n",
       "      <td>SL_D2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>80042906</td>\n",
       "      <td>chrX_100535000_100540000::chrX_99995000_100000000</td>\n",
       "      <td>SL_D2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>80042911</td>\n",
       "      <td>chrX_100590000_100595000::chrX_99995000_100000000</td>\n",
       "      <td>SL_D2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>80042912</td>\n",
       "      <td>chrX_100635000_100640000::chrX_99995000_100000000</td>\n",
       "      <td>SL_D2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>80042916</td>\n",
       "      <td>chrX_100905000_100910000::chrX_99995000_100000000</td>\n",
       "      <td>SL_D2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25300784 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  loop_name      tissue  count\n",
       "1         chr10_100000000_100005000::chr10_100015000_100...  Astrocytes      1\n",
       "16        chr10_100000000_100005000::chr10_100160000_100...  Astrocytes      1\n",
       "18        chr10_100000000_100005000::chr10_100195000_100...  Astrocytes      1\n",
       "143        chr10_10000000_10005000::chr10_11035000_11040000  Astrocytes      1\n",
       "199            chr10_1000000_1005000::chr10_1020000_1025000  Astrocytes      1\n",
       "...                                                     ...         ...    ...\n",
       "80042894  chrX_100460000_100465000::chrX_99995000_100000000       SL_D2      1\n",
       "80042906  chrX_100535000_100540000::chrX_99995000_100000000       SL_D2      1\n",
       "80042911  chrX_100590000_100595000::chrX_99995000_100000000       SL_D2      1\n",
       "80042912  chrX_100635000_100640000::chrX_99995000_100000000       SL_D2      1\n",
       "80042916  chrX_100905000_100910000::chrX_99995000_100000000       SL_D2      1\n",
       "\n",
       "[25300784 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loops_all_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get loop/anchor by tissue binary matrix, 1 or 0 depending on whether loop/anchor exists in tissue\n",
    "loops_all_df['count']=1\n",
    "loop_counts = loops_all_df.pivot_table(index='loop_name', columns='tissue',values='count', fill_value=0)\n",
    "\n",
    "anchors_all_df['count']=1\n",
    "anchor_counts = anchors_all_df.pivot_table(index='anchors', columns='tissue',values='count', fill_value=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "loop_counts_normal = loop_counts[normal_tissues]\n",
    "anchor_counts_normal = anchor_counts[normal_tissues]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "loop_counts['num_tissues'] = loop_counts.sum(axis=1)\n",
    "loop_counts_normal['num_tissues'] = loop_counts_normal.sum(axis=1)\n",
    "anchor_counts['num_tissues'] = anchor_counts.sum(axis=1)\n",
    "anchor_counts_normal['num_tissues'] = anchor_counts_normal.sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_counts_results_df = pd.DataFrame()\n",
    "all_counts_results_df['num_loops'] = loop_counts.sum(axis=0)\n",
    "all_counts_results_df['num_unique_loops'] = loop_counts[loop_counts.num_tissues==1].sum()\n",
    "all_counts_results_df['num_common_loops'] = loop_counts[loop_counts.num_tissues==(loop_counts.shape[1]-1)].sum()\n",
    "all_counts_results_df['num_anchors'] = anchor_counts.sum(axis=0)\n",
    "all_counts_results_df['num_unique_anchors'] = anchor_counts[anchor_counts.num_tissues==1].sum()\n",
    "all_counts_results_df['num_common_anchors'] = anchor_counts[anchor_counts.num_tissues==(anchor_counts.shape[1]-1)].sum()\n",
    "all_counts_results_df.drop('num_tissues',inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_counts_results_normal_df = pd.DataFrame()\n",
    "all_counts_results_normal_df['num_loops'] = loop_counts_normal.sum(axis=0)\n",
    "all_counts_results_normal_df['num_unique_loops'] = loop_counts_normal[loop_counts_normal.num_tissues==1].sum()\n",
    "all_counts_results_normal_df['num_common_loops'] = loop_counts_normal[loop_counts_normal.num_tissues==(loop_counts_normal.shape[1]-1)].sum()\n",
    "all_counts_results_normal_df['num_anchors'] = anchor_counts_normal.sum(axis=0)\n",
    "all_counts_results_normal_df['num_unique_anchors'] = anchor_counts_normal[anchor_counts_normal.num_tissues==1].sum()\n",
    "all_counts_results_normal_df['num_common_anchors'] = anchor_counts_normal[anchor_counts_normal.num_tissues==(anchor_counts_normal.shape[1]-1)].sum()\n",
    "all_counts_results_normal_df.drop('num_tissues',inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_counts_results_df.to_csv(os.path.join(save_dir, 'all_counts_results_df.csv'))\n",
    "all_counts_results_normal_df.to_csv(os.path.join(save_dir, 'all_counts_results_normal_df.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total # of loops all tissues:  25300784\n",
      "Total # unique loops all tissues:  4438651\n",
      "Total # common loops all tissues:  41329\n",
      "Total # of anchors all samples:  3211257\n",
      "Total # unique anchors all tissues:  21922\n",
      "Total # common anchors all tissues:  103354\n",
      "\n",
      "Total # of loops normal tissues :  25300784\n",
      "Total # unique loops normal tissues:  4438651\n",
      "Total # common loops normal tissues:  41329\n",
      "Total # of anchors all samples normal:  3211257\n",
      "Total # unique anchors normal tissues:  21922\n",
      "Total # common anchors normal tissues:  103354\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print('Total # of loops all tissues: ',all_counts_results_df.num_loops.sum())\n",
    "print('Total # unique loops all tissues: ',all_counts_results_df.num_unique_loops.sum())\n",
    "print('Total # common loops all tissues: ',all_counts_results_df.num_common_loops[0])\n",
    "\n",
    "print('Total # of anchors all samples: ',all_counts_results_df.num_anchors.sum())\n",
    "print('Total # unique anchors all tissues: ',all_counts_results_df.num_unique_anchors.sum())\n",
    "print('Total # common anchors all tissues: ',all_counts_results_df.num_common_anchors[0])\n",
    "\n",
    "print()\n",
    "\n",
    "print('Total # of loops normal tissues : ',all_counts_results_normal_df.num_loops.sum())\n",
    "print('Total # unique loops normal tissues: ',all_counts_results_normal_df.num_unique_loops.sum())\n",
    "print('Total # common loops normal tissues: ',all_counts_results_normal_df.num_common_loops[0])\n",
    "\n",
    "\n",
    "print('Total # of anchors all samples normal: ',all_counts_results_normal_df.num_anchors.sum())\n",
    "print('Total # unique anchors normal tissues: ',all_counts_results_normal_df.num_unique_anchors.sum())\n",
    "print('Total # common anchors normal tissues: ',all_counts_results_normal_df.num_common_anchors[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2B. save lists of anchors and loops that have been annotated \n",
    "\n",
    "- find closest gene\n",
    "- find common genes on loops "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "anchors_all = anchor_counts.index\n",
    "len(anchors_all)\n",
    "anchors_bed = pybedtools.BedTool.from_dataframe(pd.Series(anchors_all).str.split('_',expand=True)).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(541373, 541373)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(anchors_bed), len(anchors_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20314, 20314)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_df = pd.read_csv('../data/external/gene_annon_hg19.csv',index_col=0)\n",
    "tss_bed = pybedtools.BedTool.from_dataframe(tss_df[['chr','start','end', 'gene']]).sort()\n",
    "len(tss_bed), len(tss_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "anchors_bed_closest_gene = anchors_bed.closest(tss_bed, d=True)\n",
    "len(anchors_bed_closest_gene)\n",
    "anchors_closest_gene_df = anchors_bed_closest_gene.to_dataframe()\n",
    "anchors_closest_gene_df.columns = ['chr', 'start', 'stop', 'chr_tss', 'start_tss', 'stop_tss', 'gene', 'dist_to_TSS']\n",
    "anchors_closest_gene_df['anchor_name'] = anchors_closest_gene_df.chr + '_' + anchors_closest_gene_df.start.map(str) + '_' + anchors_closest_gene_df.stop.map(str)\n",
    "anchors_closest_gene_df.to_csv(os.path.join(save_dir, 'anchors_closest_gene_df.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>chr_tss</th>\n",
       "      <th>start_tss</th>\n",
       "      <th>stop_tss</th>\n",
       "      <th>gene</th>\n",
       "      <th>dist_to_TSS</th>\n",
       "      <th>anchor_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15000</td>\n",
       "      <td>20000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>49092</td>\n",
       "      <td>chr1_15000_20000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100000</td>\n",
       "      <td>105000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>134901</td>\n",
       "      <td>139379</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>29902</td>\n",
       "      <td>chr1_100000_105000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>125000</td>\n",
       "      <td>130000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>134901</td>\n",
       "      <td>139379</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>4902</td>\n",
       "      <td>chr1_125000_130000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>130000</td>\n",
       "      <td>135000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>134901</td>\n",
       "      <td>139379</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_130000_135000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>235000</td>\n",
       "      <td>240000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>134901</td>\n",
       "      <td>139379</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>95622</td>\n",
       "      <td>chr1_235000_240000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start    stop chr_tss  start_tss  stop_tss        gene  dist_to_TSS  \\\n",
       "0  chr1   15000   20000    chr1      69091     70008       OR4F5        49092   \n",
       "1  chr1  100000  105000    chr1     134901    139379  AL627309.1        29902   \n",
       "2  chr1  125000  130000    chr1     134901    139379  AL627309.1         4902   \n",
       "3  chr1  130000  135000    chr1     134901    139379  AL627309.1            0   \n",
       "4  chr1  235000  240000    chr1     134901    139379  AL627309.1        95622   \n",
       "\n",
       "          anchor_name  \n",
       "0    chr1_15000_20000  \n",
       "1  chr1_100000_105000  \n",
       "2  chr1_125000_130000  \n",
       "3  chr1_130000_135000  \n",
       "4  chr1_235000_240000  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anchors_closest_gene_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "anchors_bed_overlapping_gene = anchors_bed.intersect(tss_bed, wa=True, wb=True, f=0.5)\n",
    "len(anchors_bed_overlapping_gene)\n",
    "anchors_overlapping_gene_df = anchors_bed_overlapping_gene.to_dataframe()\n",
    "anchors_overlapping_gene_df.columns = ['chr', 'start', 'stop', 'chr_tss', 'start_tss', 'stop_tss', 'gene']\n",
    "anchors_overlapping_gene_df['anchor_name'] = anchors_overlapping_gene_df.chr + '_' + anchors_overlapping_gene_df.start.map(str) + '_' + anchors_overlapping_gene_df.stop.map(str)\n",
    "anchors_overlapping_gene_df.to_csv(os.path.join(save_dir, 'anchors_overlapping_gene_df.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>chr_tss</th>\n",
       "      <th>start_tss</th>\n",
       "      <th>stop_tss</th>\n",
       "      <th>gene</th>\n",
       "      <th>anchor_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>860000</td>\n",
       "      <td>865000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>861264</td>\n",
       "      <td>866445</td>\n",
       "      <td>AL645608.1</td>\n",
       "      <td>chr1_860000_865000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>860000</td>\n",
       "      <td>865000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>860260</td>\n",
       "      <td>879955</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>chr1_860000_865000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>865000</td>\n",
       "      <td>870000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>860260</td>\n",
       "      <td>879955</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>chr1_865000_870000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>870000</td>\n",
       "      <td>875000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>860260</td>\n",
       "      <td>879955</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>chr1_870000_875000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>875000</td>\n",
       "      <td>880000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>860260</td>\n",
       "      <td>879955</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>chr1_875000_880000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start    stop chr_tss  start_tss  stop_tss        gene  \\\n",
       "0  chr1  860000  865000    chr1     861264    866445  AL645608.1   \n",
       "1  chr1  860000  865000    chr1     860260    879955      SAMD11   \n",
       "2  chr1  865000  870000    chr1     860260    879955      SAMD11   \n",
       "3  chr1  870000  875000    chr1     860260    879955      SAMD11   \n",
       "4  chr1  875000  880000    chr1     860260    879955      SAMD11   \n",
       "\n",
       "          anchor_name  \n",
       "0  chr1_860000_865000  \n",
       "1  chr1_860000_865000  \n",
       "2  chr1_865000_870000  \n",
       "3  chr1_870000_875000  \n",
       "4  chr1_875000_880000  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anchors_overlapping_gene_df[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2C . common loops annotated with genes\n",
    "\n",
    "#### all tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(41329, 2)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_loops_df = pd.Series(loop_counts[loop_counts.num_tissues==(loop_counts.shape[1]-1)].index).str.split('::', expand=True)\n",
    "common_loops_df.columns = ['source', 'target']\n",
    "common_loops_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(37987, 2)\n",
      "(24207, 2)\n",
      "9584\n"
     ]
    }
   ],
   "source": [
    "common_loop_anchors_df = pd.DataFrame({'anchor_name':common_loops_df.stack().reset_index(drop=True).drop_duplicates()})\n",
    "common_loop_anchors_df = common_loop_anchors_df.merge(anchors_overlapping_gene_df[['anchor_name', 'gene']], how='left',on='anchor_name')\n",
    "print(common_loop_anchors_df.shape)\n",
    "common_loop_anchors_df= common_loop_anchors_df.dropna()\n",
    "print(common_loop_anchors_df.shape)\n",
    "common_loop_genes = sorted(common_loop_anchors_df.gene.unique())\n",
    "print(len(common_loop_genes))\n",
    "common_loop_genes_file = os.path.join(save_dir, 'common_loop_genes.txt')\n",
    "with open(common_loop_genes_file, 'w') as f:\n",
    "    for gene in common_loop_genes:\n",
    "        f.write(gene + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### normal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(41329, 2)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_loops_df_normal = pd.Series(loop_counts_normal[loop_counts_normal.num_tissues==(loop_counts_normal.shape[1]-1)].index).str.split('::', expand=True)\n",
    "common_loops_df_normal.columns = ['source', 'target']\n",
    "common_loops_df_normal.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(37987, 2)\n",
      "(24207, 2)\n",
      "9584\n"
     ]
    }
   ],
   "source": [
    "common_loop_anchors_normal_df = pd.DataFrame({'anchor_name':common_loops_df_normal.stack().reset_index(drop=True).drop_duplicates()})\n",
    "common_loop_anchors_normal_df = common_loop_anchors_normal_df.merge(anchors_overlapping_gene_df[['anchor_name', 'gene']], how='left',on='anchor_name')\n",
    "print(common_loop_anchors_normal_df.shape)\n",
    "common_loop_anchors_normal_df= common_loop_anchors_normal_df.dropna()\n",
    "print(common_loop_anchors_normal_df.shape)\n",
    "common_loop_normal_genes = sorted(common_loop_anchors_normal_df.gene.unique())\n",
    "print(len(common_loop_normal_genes))\n",
    "common_loop_normal_genes_file = os.path.join(save_dir, 'common_loop_normal_genes.txt')\n",
    "with open(common_loop_normal_genes_file, 'w') as f:\n",
    "    for gene in common_loop_normal_genes:\n",
    "        f.write(gene + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. get per tissue information on anchors\n",
    "- % PP\t: \n",
    "- % P-PIR\t: PIR = not accessible\n",
    "- % PIR-PIR\n",
    "- % ACCESSIBLE\n",
    "- % P-P+ ACC\n",
    "- % EP + ACC\n",
    "- % EE + ACC\n",
    "- TOTAL %\n",
    "\n",
    "Promoter/Enhancer Annotations of HiChIP Loops\n",
    "Promoters were defined as 1 kb regions centered at the TSS, and enhancers\n",
    "were identified as chromHMM enhancers not overlapping with promoters in any\n",
    "cell type. We annotated loop anchors as ‘other’ if the anchor did not contain a\n",
    "promoter or enhancer as defined above\n",
    "\n",
    "\n",
    "we define: i modified our stuff\n",
    "P-P (don't care about accessibility here\n",
    "9:22\n",
    "P-E = P-PIRacc (so a promoter region connecting to an accessible region that is not a promoter)\n",
    "9:22\n",
    "E-E = PIRacc-PIRacc\n",
    "\n",
    "Note: our promoter definition is slightly different than what max is our is 2kb upstream, +500 bp downstream\n",
    "\n",
    "\n",
    "05/12/2020 - change, how we get promoter regions, to look at exon regions, 1/2 of exon must overlap with some of the anchor region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1196293"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_bed = pybedtools.BedTool('../data/external/gencode.v19.exon.bed')\n",
    "len(exon_bed)# 1196293"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Astrocytes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H9_D0\n",
      "H9_D10\n",
      "H9_D2\n",
      "H9_D28\n",
      "SLC_D0\n",
      "SLC_D2\n",
      "SL_D0\n",
      "SL_D2\n",
      "CPU times: user 1min 23s, sys: 2.94 s, total: 1min 26s\n",
      "Wall time: 40.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "perc_anchor_accessible_dict = {}\n",
    "\n",
    "for loop_file in sorted(loop_files):\n",
    "    loop_filename = os.path.basename(loop_file)\n",
    "    tissue = loop_filename.split('.')[0]\n",
    "    print(tissue)\n",
    "    loop_df = pd.read_csv(loop_file, index_col=0)\n",
    "    anchor_file = os.path.join(anchor_dir, tissue+'.anchors.csv')\n",
    "    anchor_df = pd.read_csv(anchor_file, index_col=0)\n",
    "    anchor_df = anchor_df[['chr','start','end','anchors','sample']]#[:5]\n",
    "    anchor_bed = pybedtools.BedTool.from_dataframe(anchor_df) #101559\n",
    "\n",
    "    anchor_annon_file = os.path.join(anchor_annon_dir, tissue+'_annon.bed')\n",
    "    if not os.path.exists(anchor_annon_file):\n",
    "        print(tissue, 'not anchor annon file.. skipping')\n",
    "        continue\n",
    "    anchor_annon_df = pd.read_csv(anchor_annon_file, names=['chr', 'start', 'stop', 'name', 'atac_chr', 'atac_start', 'atac_stop'], sep='\\t')\n",
    "    anchor_to_count_dict = anchor_annon_df.groupby('name').name.count().to_dict()\n",
    "    anchors_acc = anchor_to_count_dict.keys()\n",
    "    \n",
    "    \n",
    "    promoter_anchor_file = os.path.join(promoter_annon_dir, 'promoter_'+tissue+'_annon.bed')\n",
    "    if not os.path.exists(promoter_anchor_file):\n",
    "        print(tissue, 'not promoter annon file.. skipping')\n",
    "        continue\n",
    "    promoter_anchor_df = pd.read_csv(promoter_anchor_file, names=['chr', 'start', 'stop', 'tss', 'anchor_chr', 'anchor_start', \n",
    "                                                             'anchor_stop', 'anchor', 'overlap'], sep='\\t')\n",
    "    promoter_anchors = promoter_anchor_df.anchor.unique()\n",
    "#     #### PART ADDED FOR EXOM #####\n",
    "#     anchor_bed_exons = anchor_bed.intersect(exon_bed,wa=True,wb=True,F=.5)# 1/2 of exon must overlap with some of the anchor region\n",
    "#     exon_anchors = anchor_bed_exons.to_dataframe().name.unique()\n",
    "#     promoter_anchors = np.union1d(promoter_anchors,exon_anchors)\n",
    "    \n",
    "#     #####\n",
    "    \n",
    "    anchor_df['is_P'] = anchor_df['anchors'].isin(promoter_anchors)\n",
    "    anchor_df['is_acc'] = anchor_df['anchors'].isin(anchors_acc)\n",
    "    anchor_df['is_Pacc'] = anchor_df['is_P'] & anchor_df['is_acc']\n",
    "    anchor_df['is_PIRacc'] = (~anchor_df['is_P']) & anchor_df['is_acc']\n",
    "    \n",
    "    \n",
    "    loop_df['source_P'] = loop_df.source.isin(promoter_anchors)\n",
    "    loop_df['target_P'] = loop_df.target.isin(promoter_anchors)\n",
    "    \n",
    "    loop_df['source_acc'] = loop_df.source.isin(anchors_acc)\n",
    "    loop_df['target_acc'] = loop_df.target.isin(anchors_acc)   \n",
    "    \n",
    "\n",
    "    loop_df['source_acc_P'] = loop_df['source_acc'] & loop_df['source_P']\n",
    "    loop_df['target_acc_P'] = loop_df['target_acc'] & loop_df['target_P']\n",
    "    loop_df['source_acc_PIR'] = loop_df['source_acc'] & (~loop_df['source_P'])\n",
    "    loop_df['target_acc_PIR'] = loop_df['target_acc'] & (~loop_df['target_P'])\n",
    "    \n",
    "    \n",
    "    loop_df['is_P_P'] = loop_df['source_P'] & loop_df['target_P']\n",
    "    loop_df['is_P_PIR'] = loop_df['source_P'] ^ loop_df['target_P']\n",
    "    loop_df['is_PIR_PIR'] = ~(loop_df['source_P'] | loop_df['target_P'])\n",
    "    \n",
    "    loop_df['both_anchor_acc'] = loop_df['source_acc'] & loop_df['target_acc']\n",
    "    loop_df['one_anchor_acc'] = loop_df['source_acc'] ^ loop_df['target_acc']\n",
    "    loop_df['both_anchor_no_acc'] = ~(loop_df['source_acc'] | loop_df['target_acc'])\n",
    "  \n",
    "    loop_df['is_P_PIRacc'] = loop_df['is_P_PIR'] & (loop_df['source_acc_PIR'] | loop_df['target_acc_PIR']) \n",
    "\n",
    "    \n",
    "    loop_df['is_Pacc_Pacc'] = loop_df['is_P_P']  & loop_df['both_anchor_acc'] \n",
    "    loop_df['is_Pacc_PIRacc'] = loop_df['is_P_PIR']  & loop_df['both_anchor_acc'] \n",
    "    loop_df['is_PIRacc_PIRacc'] = loop_df['is_PIR_PIR'] & loop_df['both_anchor_acc'] \n",
    "    \n",
    "  \n",
    "    ## SUMMATIVE STATS PER TISSUE ##\n",
    "    num_anchor = anchor_df.shape[0]\n",
    "    perc_anchor_accessible = len(anchors_acc)/num_anchor\n",
    "    perc_anchor_P_or_PIRacc = (anchor_df['is_PIRacc'].sum() + len(promoter_anchors))/num_anchor\n",
    "    perc_anchor_promoter = len(promoter_anchors)/num_anchor\n",
    "#     perc_anchor_Pacc = anchor_df['is_Pacc'].sum()/num_anchor\n",
    "#     perc_anchor_PIR = 1 - len(promoter_anchors)/num_anchor\n",
    "    perc_anchor_PIRacc = anchor_df['is_PIRacc'].sum()/num_anchor\n",
    "\n",
    "    num_loop = loop_df.shape[0]\n",
    "    perc_both_anchor_acc = loop_df['both_anchor_acc'].sum()/num_loop\n",
    "    perc_one_anchor_acc = loop_df['one_anchor_acc'].sum()/num_loop\n",
    "\n",
    "    perc_is_P_P = loop_df['is_P_P'].sum()/num_loop\n",
    "    perc_is_P_PIR = loop_df['is_P_PIR'].sum()/num_loop\n",
    "    perc_is_PIR_PIR = loop_df['is_PIR_PIR'].sum()/num_loop\n",
    "    perc_is_P_PIRacc = loop_df['is_P_PIRacc'].sum()/num_loop\n",
    "    perc_is_PIRacc_PIRacc = loop_df['is_PIRacc_PIRacc'].sum()/num_loop\n",
    "    \n",
    "    \n",
    "    ### modification\n",
    "    num_loops_considered = loop_df['is_P_P'].sum()+loop_df['is_P_PIRacc'].sum()+loop_df['is_PIRacc_PIRacc'].sum()\n",
    "    perc_is_P_P_adj = loop_df['is_P_P'].sum()/num_loops_considered\n",
    "    perc_is_P_PIRacc_adj = loop_df['is_P_PIRacc'].sum()/num_loops_considered\n",
    "    perc_is_PIRacc_PIRacc_adj = loop_df['is_PIRacc_PIRacc'].sum()/num_loops_considered\n",
    "     \n",
    "#     num_both_anchor_acc = loop_df['both_anchor_acc'].sum()\n",
    "#     perc_is_Pacc_Pacc_of_both_acc = loop_df['is_Pacc_Pacc'].sum()/num_both_anchor_acc\n",
    "#     perc_is_Pacc_PIRacc_of_both_acc = loop_df['is_Pacc_PIRacc'].sum()/num_both_anchor_acc\n",
    "#     perc_is_PIRacc_PIRacc_of_both_acc = loop_df['is_PIRacc_PIRacc'].sum()/num_both_anchor_acc\n",
    "    \n",
    "    perc_anchor_accessible_dict[tissue] = {'num_anchor':num_anchor,\n",
    "                                           'perc_anchor_P_or_PIRacc':perc_anchor_P_or_PIRacc,\n",
    "                                           'perc_anchor_accessible':perc_anchor_accessible,\n",
    "                                           'perc_anchor_promoter':perc_anchor_promoter,\n",
    "#                                            'perc_anchor_Pacc':perc_anchor_Pacc,\n",
    "#                                            'perc_anchor_PIR':perc_anchor_PIR,\n",
    "                                           'perc_anchor_PIRacc':perc_anchor_PIRacc,\n",
    "                                           'empty1':'',\n",
    "                                           \n",
    "                                           'num_loop':num_loop, \n",
    "                                           'perc_both_anchor_acc':perc_both_anchor_acc,\n",
    "                                           'perc_one_anchor_acc':perc_one_anchor_acc,\n",
    "                                           'perc_is_P_P':perc_is_P_P,\n",
    "                                           'perc_is_P_PIR':perc_is_P_PIR,\n",
    "                                           'perc_is_PIR_PIR':perc_is_PIR_PIR,\n",
    "                                           'perc_is_P_PIRacc':perc_is_P_PIRacc,\n",
    "                                           'perc_is_PIRacc_PIRacc':perc_is_PIRacc_PIRacc  ,\n",
    "                                           'empty2':'',\n",
    "                                           \n",
    "                                           'num_loops_considered':num_loops_considered,\n",
    "                                           'perc_is_P_P_adj':perc_is_P_P_adj,\n",
    "                                           'perc_is_P_PIRacc_adj':perc_is_P_PIRacc_adj,\n",
    "                                           'perc_is_PIR_PIRacc_adj':perc_is_PIRacc_PIRacc_adj,\n",
    "                                           \n",
    "    }\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "perc_anchor_accessible_df = pd.DataFrame.from_dict(perc_anchor_accessible_dict,orient='index')\n",
    "perc_anchor_accessible_df.to_csv(os.path.join(save_dir, 'perc_anchor_accessible_df_updated.csv'))\n",
    "perc_anchor_accessible_normal_df = perc_anchor_accessible_df[perc_anchor_accessible_df.index.isin(normal_tissues)]\n",
    "perc_anchor_accessible_normal_df.to_csv(os.path.join(save_dir, 'perc_anchor_accessible_normal_df_updated.csv'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_anchor</th>\n",
       "      <th>perc_anchor_P_or_PIRacc</th>\n",
       "      <th>perc_anchor_accessible</th>\n",
       "      <th>perc_anchor_promoter</th>\n",
       "      <th>perc_anchor_PIRacc</th>\n",
       "      <th>empty1</th>\n",
       "      <th>num_loop</th>\n",
       "      <th>perc_both_anchor_acc</th>\n",
       "      <th>perc_one_anchor_acc</th>\n",
       "      <th>perc_is_P_P</th>\n",
       "      <th>perc_is_P_PIR</th>\n",
       "      <th>perc_is_PIR_PIR</th>\n",
       "      <th>perc_is_P_PIRacc</th>\n",
       "      <th>perc_is_PIRacc_PIRacc</th>\n",
       "      <th>empty2</th>\n",
       "      <th>num_loops_considered</th>\n",
       "      <th>perc_is_P_P_adj</th>\n",
       "      <th>perc_is_P_PIRacc_adj</th>\n",
       "      <th>perc_is_PIR_PIRacc_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>322550</td>\n",
       "      <td>0.147146</td>\n",
       "      <td>0.050460</td>\n",
       "      <td>0.106678</td>\n",
       "      <td>0.040468</td>\n",
       "      <td></td>\n",
       "      <td>1839205</td>\n",
       "      <td>0.015882</td>\n",
       "      <td>0.179444</td>\n",
       "      <td>0.125717</td>\n",
       "      <td>0.874283</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.084548</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>386721</td>\n",
       "      <td>0.597899</td>\n",
       "      <td>0.402101</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H9_D0</td>\n",
       "      <td>306851</td>\n",
       "      <td>0.205960</td>\n",
       "      <td>0.110480</td>\n",
       "      <td>0.114297</td>\n",
       "      <td>0.091663</td>\n",
       "      <td></td>\n",
       "      <td>1318716</td>\n",
       "      <td>0.034560</td>\n",
       "      <td>0.265318</td>\n",
       "      <td>0.141899</td>\n",
       "      <td>0.858101</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.131234</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>360185</td>\n",
       "      <td>0.519525</td>\n",
       "      <td>0.480475</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H9_D10</td>\n",
       "      <td>230431</td>\n",
       "      <td>0.204998</td>\n",
       "      <td>0.085327</td>\n",
       "      <td>0.136297</td>\n",
       "      <td>0.068702</td>\n",
       "      <td></td>\n",
       "      <td>712533</td>\n",
       "      <td>0.030627</td>\n",
       "      <td>0.234692</td>\n",
       "      <td>0.154784</td>\n",
       "      <td>0.845216</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.119785</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>195640</td>\n",
       "      <td>0.563734</td>\n",
       "      <td>0.436266</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H9_D2</td>\n",
       "      <td>461684</td>\n",
       "      <td>0.180333</td>\n",
       "      <td>0.103941</td>\n",
       "      <td>0.088992</td>\n",
       "      <td>0.091342</td>\n",
       "      <td></td>\n",
       "      <td>6720106</td>\n",
       "      <td>0.017614</td>\n",
       "      <td>0.213867</td>\n",
       "      <td>0.108636</td>\n",
       "      <td>0.891364</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.103185</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>1423457</td>\n",
       "      <td>0.512867</td>\n",
       "      <td>0.487133</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H9_D28</td>\n",
       "      <td>345113</td>\n",
       "      <td>0.161860</td>\n",
       "      <td>0.044913</td>\n",
       "      <td>0.126411</td>\n",
       "      <td>0.035449</td>\n",
       "      <td></td>\n",
       "      <td>1256959</td>\n",
       "      <td>0.009152</td>\n",
       "      <td>0.130480</td>\n",
       "      <td>0.137398</td>\n",
       "      <td>0.862602</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.060738</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>249049</td>\n",
       "      <td>0.693454</td>\n",
       "      <td>0.306546</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SLC_D0</td>\n",
       "      <td>348166</td>\n",
       "      <td>0.225967</td>\n",
       "      <td>0.140669</td>\n",
       "      <td>0.106392</td>\n",
       "      <td>0.119575</td>\n",
       "      <td></td>\n",
       "      <td>2659082</td>\n",
       "      <td>0.055946</td>\n",
       "      <td>0.328058</td>\n",
       "      <td>0.138145</td>\n",
       "      <td>0.861855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.177202</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>838532</td>\n",
       "      <td>0.438073</td>\n",
       "      <td>0.561927</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SLC_D2</td>\n",
       "      <td>438625</td>\n",
       "      <td>0.213757</td>\n",
       "      <td>0.140115</td>\n",
       "      <td>0.090955</td>\n",
       "      <td>0.122802</td>\n",
       "      <td></td>\n",
       "      <td>4282155</td>\n",
       "      <td>0.037083</td>\n",
       "      <td>0.277877</td>\n",
       "      <td>0.110521</td>\n",
       "      <td>0.889479</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.151784</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>1123231</td>\n",
       "      <td>0.421344</td>\n",
       "      <td>0.578656</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SL_D0</td>\n",
       "      <td>297076</td>\n",
       "      <td>0.235781</td>\n",
       "      <td>0.142388</td>\n",
       "      <td>0.115492</td>\n",
       "      <td>0.120289</td>\n",
       "      <td></td>\n",
       "      <td>1187599</td>\n",
       "      <td>0.052260</td>\n",
       "      <td>0.314103</td>\n",
       "      <td>0.140711</td>\n",
       "      <td>0.859289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.171048</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>370245</td>\n",
       "      <td>0.451344</td>\n",
       "      <td>0.548656</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SL_D2</td>\n",
       "      <td>460761</td>\n",
       "      <td>0.189020</td>\n",
       "      <td>0.114793</td>\n",
       "      <td>0.088951</td>\n",
       "      <td>0.100069</td>\n",
       "      <td></td>\n",
       "      <td>5324429</td>\n",
       "      <td>0.024244</td>\n",
       "      <td>0.236247</td>\n",
       "      <td>0.106397</td>\n",
       "      <td>0.893603</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.120768</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>1209523</td>\n",
       "      <td>0.468368</td>\n",
       "      <td>0.531632</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            num_anchor  perc_anchor_P_or_PIRacc  perc_anchor_accessible  \\\n",
       "Astrocytes      322550                 0.147146                0.050460   \n",
       "H9_D0           306851                 0.205960                0.110480   \n",
       "H9_D10          230431                 0.204998                0.085327   \n",
       "H9_D2           461684                 0.180333                0.103941   \n",
       "H9_D28          345113                 0.161860                0.044913   \n",
       "SLC_D0          348166                 0.225967                0.140669   \n",
       "SLC_D2          438625                 0.213757                0.140115   \n",
       "SL_D0           297076                 0.235781                0.142388   \n",
       "SL_D2           460761                 0.189020                0.114793   \n",
       "\n",
       "            perc_anchor_promoter  perc_anchor_PIRacc empty1  num_loop  \\\n",
       "Astrocytes              0.106678            0.040468          1839205   \n",
       "H9_D0                   0.114297            0.091663          1318716   \n",
       "H9_D10                  0.136297            0.068702           712533   \n",
       "H9_D2                   0.088992            0.091342          6720106   \n",
       "H9_D28                  0.126411            0.035449          1256959   \n",
       "SLC_D0                  0.106392            0.119575          2659082   \n",
       "SLC_D2                  0.090955            0.122802          4282155   \n",
       "SL_D0                   0.115492            0.120289          1187599   \n",
       "SL_D2                   0.088951            0.100069          5324429   \n",
       "\n",
       "            perc_both_anchor_acc  perc_one_anchor_acc  perc_is_P_P  \\\n",
       "Astrocytes              0.015882             0.179444     0.125717   \n",
       "H9_D0                   0.034560             0.265318     0.141899   \n",
       "H9_D10                  0.030627             0.234692     0.154784   \n",
       "H9_D2                   0.017614             0.213867     0.108636   \n",
       "H9_D28                  0.009152             0.130480     0.137398   \n",
       "SLC_D0                  0.055946             0.328058     0.138145   \n",
       "SLC_D2                  0.037083             0.277877     0.110521   \n",
       "SL_D0                   0.052260             0.314103     0.140711   \n",
       "SL_D2                   0.024244             0.236247     0.106397   \n",
       "\n",
       "            perc_is_P_PIR  perc_is_PIR_PIR  perc_is_P_PIRacc  \\\n",
       "Astrocytes       0.874283              0.0          0.084548   \n",
       "H9_D0            0.858101              0.0          0.131234   \n",
       "H9_D10           0.845216              0.0          0.119785   \n",
       "H9_D2            0.891364              0.0          0.103185   \n",
       "H9_D28           0.862602              0.0          0.060738   \n",
       "SLC_D0           0.861855              0.0          0.177202   \n",
       "SLC_D2           0.889479              0.0          0.151784   \n",
       "SL_D0            0.859289              0.0          0.171048   \n",
       "SL_D2            0.893603              0.0          0.120768   \n",
       "\n",
       "            perc_is_PIRacc_PIRacc empty2  num_loops_considered  \\\n",
       "Astrocytes                    0.0                       386721   \n",
       "H9_D0                         0.0                       360185   \n",
       "H9_D10                        0.0                       195640   \n",
       "H9_D2                         0.0                      1423457   \n",
       "H9_D28                        0.0                       249049   \n",
       "SLC_D0                        0.0                       838532   \n",
       "SLC_D2                        0.0                      1123231   \n",
       "SL_D0                         0.0                       370245   \n",
       "SL_D2                         0.0                      1209523   \n",
       "\n",
       "            perc_is_P_P_adj  perc_is_P_PIRacc_adj  perc_is_PIR_PIRacc_adj  \n",
       "Astrocytes         0.597899              0.402101                     0.0  \n",
       "H9_D0              0.519525              0.480475                     0.0  \n",
       "H9_D10             0.563734              0.436266                     0.0  \n",
       "H9_D2              0.512867              0.487133                     0.0  \n",
       "H9_D28             0.693454              0.306546                     0.0  \n",
       "SLC_D0             0.438073              0.561927                     0.0  \n",
       "SLC_D2             0.421344              0.578656                     0.0  \n",
       "SL_D0              0.451344              0.548656                     0.0  \n",
       "SL_D2              0.468368              0.531632                     0.0  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "perc_anchor_accessible_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.12935646733007433 0.8706435326699258 0.0\n"
     ]
    }
   ],
   "source": [
    "print(perc_anchor_accessible_df.perc_is_P_P.mean(),\n",
    "      perc_anchor_accessible_df.perc_is_P_PIR.mean(),\n",
    "      perc_anchor_accessible_df.perc_is_PIR_PIR.mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.518512021120676 0.48148797887932404 0.0\n"
     ]
    }
   ],
   "source": [
    "print(perc_anchor_accessible_df.perc_is_P_P_adj\t.mean(),\n",
    "      perc_anchor_accessible_df.perc_is_P_PIRacc_adj.mean(),\n",
    "      perc_anchor_accessible_df.perc_is_PIR_PIRacc_adj.mean())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "means across the board: P-P 51.9%, E-P 48.1%, E-E 0%, where P = (in promoter region OR >1/2 exon region overlaps with anchor region) AND accessible, and E = not in promoter/exon regions AND accessible (edited) \n",
    "6:38\n",
    "namyoung said: E-E, HiChIP, 40%,  E-P 40%, P-P% 20%\n",
    "6:38\n",
    "rip\n",
    "6:40\n",
    "if you don't care about accessibility: percentages are\n",
    "P-P: 30.3%, P-PIR: 42.0%, PIR-PIR: 27.%"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Maxs paper\"Promoter/Enhancer Annotations of HiChIP Loops Promoters were defined as 1 kb regions centered at the TSS, and enhancers were identified as chromHMM enhancers not overlapping with promoters in any cell type. We annotated loop anchors as ‘other’ if the anchor did not contain a promoter or enhancer as defined above.\"\\"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3B:  replicate Song 4C, figure out \n",
    "\n",
    "figure out counts for the types as outlined by Song, et al fig 4b/c ish \n",
    "Types (enhancer = has atac region)\n",
    "- I. Enhancer interacts with distal gene(s) only  \n",
    "- II. Enhancer interacts with nearest and distal gene(s)  \n",
    "- III. Enhancer interacts with nearest gene only \n",
    "- IV. Enhancer cannot be resolved for interactions with nearest gene\n",
    "\n",
    "calculate for loops that are P-PIRacc and for Pacc-PIRacc\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count         9.000000\n",
       "mean     350109.555556\n",
       "std      260790.287280\n",
       "min       76345.000000\n",
       "25%      155501.000000\n",
       "50%      203137.000000\n",
       "75%      643021.000000\n",
       "max      693413.000000\n",
       "Name: num_is_P_PIRacc, dtype: float64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numbers_P_PIRacc = perc_anchor_accessible_df[['num_loop','perc_is_P_PIRacc']]\n",
    "numbers_P_PIRacc['num_is_P_PIRacc'] = (numbers_P_PIRacc.num_loop * numbers_P_PIRacc.perc_is_P_PIRacc).map(int)\n",
    "numbers_P_PIRacc.num_is_P_PIRacc.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fc92e5532d0>"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZkAAAHhCAYAAAC81MNiAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3df5ieZX3n/fcHIkoKAoGIKCGhFPmxCAlMpWIxgHZ1uxxGClasyo/6PGwtHnbr0keRyvbQw6qV3W0LtStaUajPiuahha0/wCMScRHECUlAjAKKIYiFAIJQskDg+/xxX+mOycxkJjPn3DPJ+3Uc9zHXj/M6r++ZTPKZ87ruue5UFZIktbBTvwuQJG2/DBlJUjOGjCSpGUNGktSMISNJasaQkSQ1M6vfBUwn++yzTy1YsKDfZUjSjLJixYqHqmrucPsMmSEWLFjA4OBgv8uQpBklydqR9nm5TJLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNdMsZJI8sdn6WUku6ZbnJ1mW5LYky5PsP0o/C5JsSLIyyZoktyQ5c8j+t3b93Jbk20mOGrLvj5PckeR7Sf5Hkhe0GKskaXj9mslcBFxeVUcCHwQ+spX2P6qqRVV1GHA68MdJzu723QMs7vr6EHApQJKXAu8GBqrqCGDn7lhJ0hTpV8gcDizrlq8Hloz1wKr6MfAeegFCVX27qn7e7b4ZGDormgXsmmQWMBu4f4J1S5LGoWXI7Jpk1aYXvRnLJquBU7vlU4Ddk+w9jr5vBQ4dZvs7gK8CVNVP6c2Y7gV+BjxWVddtfkCSc5IMJhlcv379OEqQJG1Ny5DZUFULN72AC4fsOw9YnGQlsBj4KbBxHH1niw3JifRC5r3d+l70ZkgHAi8BfiXJ2zY/rqouraqBqhqYO3fuOEqYXPc+/CS/+8mb+LX3f4Xf/eRN3Pvwk32rRZImS18ul1XV/VX1O1W1CLig2/bYOLpYBKzZtJLkSODTwJKqerjb/FrgnqpaX1XPAFcBx03KABo4b+lqbrnnETY+V9xyzyOct3R1v0uSpAnrS8gk2SfJpnOfD3xmHMcuoHcZ7OJu/QB6AfL2qrpzSNN7gd9IMjtJgNcwJJimm1vX/nzUdUmaifp14/8E4IdJ7gT2BT68lfYHbXoLM/BF4OKquqzbdyGwN/CJ7v7PIEBVfQdYSu/+ze30xnrppI9kkhw9f69R1yVpJkpV9buGaWNgYKAGBwf7cu57H36S85au5ta1P+fo+Xtx0WlHccDes/tSiySNR5IVVTUw3L5ZU12MhnfA3rP54n94Zb/LkKRJNW1CJsnLgSs22/xUVR3bj3okSRM3bUKmqm4HFva7DknS5PEBmZKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUzJhCJskpSSrJoVtp9/7JKWvUc7wxyeGtzyNJmrixzmTeAvwv4PSttBs2ZNIzWbOmNwKGzEz1yD1w2W/DB/fufX3knn5XJKmhrf7Hn2Q34FXAO+hCJsl+SW5IsirJ95Icn+SjwK7dts8nWZBkTZJPALcC85K8Jcnt3TEfG3KO1ye5NcnqJMuS7JTkriRzu/07Jbk7yauBNwAf785zUPf6WpIVSb61abaV5E3deVYnuWGy/+C0ja4+F9beCM9t7H29+tx+VySpoVljaPNG4GtVdWeSR5IcDZwIXFtVH06yMzC7qr6V5F1VtRAgyQLgEODsqvrDJC8BPgYcA/wcuC7JG4EbgU8Br66qe5LMqarnkvw98FbgL4HXAqur6oYk1wD/VFVLu/MsA/6gqu5KcizwCeAk4ELgdVX10yR7jjS4JOcA5wAccMAB4/ij0zZZ953R1yVtV8ZyCestwBe65S90698Fzk7yZ8DLq+rxEY5dW1U3d8u/DiyvqvVVtRH4PPBq4DeAG6rqHoCqeqRr/xngjG7594HLNu+8m2UdB3wpySrgk8B+3e4bgc8m+b+BnUcaXFVdWlUDVTUwd+7cUf4YNCnmHTv6uqTtyqghk2RverOCTyf5CfAnwJuBb9ELiJ8CVyQ5Y4Qu/mVodyOdBqjNN1bVOuCBJCcBxwJfHaH+R6tq4ZDXYd3xfwD8KTAPWNWNRf225G9g/qtgp1m9r0v+pt8VSWpoazOZ04DLq2p+VS2oqnnAPfQC5sGq+hTwd8DRXftnkjxvhL6+AyxOsk93ie0twDeBm7rtBwIkmTPkmE8Dfw98saqe7bY9DuwOUFW/AO5J8qbu2CQ5qls+qKq+U1UXAg/RCxv125wD4eyvwIUP977OObDfFUlqaGsh8xbgHzbb9v8Bn6U3O1gJnAr8VbfvUuC2JJ/fvKOq+hlwPnA9sBq4taqurqr19O6JXJVkNXDlkMOuAXbjly+VfQH4kyQrkxxE777NO7pj7wCWdO0+vulNBsAN3TklSVMoVVtcqZo2kgwA/62qjp+K8w0MDNTg4OBUnEqSthtJVlTVwHD7xvLusr5I8j7gnfRmKpKkGWjaPlamqj7a3Qv6X/2uRZK0baZtyEiSZj5DRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISGOw7vF1nPW1s1h0+SLO+tpZrHt8Xb9LkmYEQ0Yagw/c+AFWPLCCjbWRFQ+s4AM3fqDfJUkzwpSHTJInNls/K8kl3fL8JMuS3JZkeZL9R+lnQZINSVYmWZPkliRnDtn/1q6f25J8O8lR7Ual7d3qB1ePui5peNNtJnMRcHlVHQl8EPjIVtr/qKoWVdVhwOnAHyc5u9t3D7C46+tDwKWtitb276gXHTXquqThTbeQORxY1i1fDywZ64FV9WPgPcC7u/VvV9XPu903AyPOiqSt+dCrPsQx+x7DrMzimH2P4UOv+lC/S5JmhFl9OOeuSVYNWZ8DXNMtrwZOBf4KOAXYPcneVfXwGPu+FTh0mO3vAL463AFJzgHOATjggAPGeBrtaObtPo/Pvv6z/S5DmnH6MZPZUFULN72AC4fsOw9YnGQlsBj4KbBxHH1niw3JifRC5r3DHVBVl1bVQFUNzJ07dxynkiRtTT9mMiOqqvuB3wFIshtwalU9No4uFgFrNq0kORL4NPDvxjEbkiRNkml1TybJPkk21XQ+8JlxHLuA3hsHLu7WDwCuAt5eVXdObqWSpLGYVjMZ4ATgI0kKuAE4dyvtD+ourb0AeBy4uKou6/ZdCOwNfCIJwMaqGmhStSRpWKmqftcwbQwMDNTg4GC/y5CkGSXJipF+iJ9Wl8skSduX6Xa5bAtJXg5csdnmp6rq2H7UI0kau2kfMlV1O7Cw33VIksbPy2WSpGYMGUlSM4aMJKkZQ0aS1IwhI0lqxpCRJDVjyEiSmjFkJEnNGDKSpGYMGUlSM4aMJKkZQ0aS1IwhI0lqxpCRJDVjyEiSmjFkJEnNGDKSpGYMGUlSM4aMJKkZQ0aS1IwhI0lqxpCRJDVjyEiSmjFkJEnNGDKSpGYMGUlSM4aMJKkZQ0aS1IwhI0lqxpCRJDVjyEiSmjFkJEnNGDKSpGYMGUlSM4aMJKkZQ0aS1EyzkEnyxGbrZyW5pFuen2RZktuSLE+y/yj9LEiyIcnKJGuS3JLkzCH7D01yU5Knkpy32bGvT/LDJHcned9kj1GSNLpZfTrvRcDlVfW5JCcBHwHePkr7H1XVIoAkvwpclWSnqroMeAR4N/DGoQck2Rn4G+C3gPuA7ya5pqq+P/nDkSQNp1+Xyw4HlnXL1wNLxnpgVf0YeA+9YKGqHqyq7wLPbNb0FcDdVfXjqnoa+MJ4ziNJmriWIbNrklWbXsAHh+xbDZzaLZ8C7J5k73H0fStw6FbavBRYN2T9vm7bL0lyTpLBJIPr168fRwnS9PH0unWsfdvbWXPEy1n7trfz9Lp1Wz9ImgItQ2ZDVS3c9AIuHLLvPGBxkpXAYuCnwMZx9J1tbFNbbKi6tKoGqmpg7ty54yhBmj5+dv77eXJwEDZu5MnBQX52/vv7XZIE9OmeTFXdD/wOQJLdgFOr6rFxdLEIWLOVNvcB84as7w/cP546pZniyVWrRl2X+qUv92SS7JNk07nPBz4zjmMX0HvjwMVbafpd4OAkBybZBTgduGb81UrT3+yFC0ddl/qlXzf+TwB+mOROYF/gw1tpf9CmtzADXwQu7t5ZRpIXJ7mP3psB/jTJfUleWFUbgXcB19Kb9Xyxqu5oNB6pr/b7yJ8ze2AAZs1i9sAA+33kz/tdkgRAqra4TbHDGhgYqMHBwX6XIUkzSpIVVTUw3D5/41+S1Ey/fhlzC0leDlyx2eanqurYftQjSZq4aRMyVXU74N1KSdqOeLlMktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpqZ1e8CJE2ux9Zv4BuXr+Gff/QYLz5oD0464zD2mLtrv8vSDqqvM5kkFyS5I8ltSVYlOTbJ8iQDYzj2hCSPJVmZ5IdJbkhy8pD970ny/a7vZUnmtx2NND184/I13H/Xozz3XHH/XY/yjcvX9Lsk7cD6NpNJ8krgZODoqnoqyT7ALuPs5ltVdXLX30LgH5NsqKplwEpgoKqeTPJO4C+AN0/iEKRp6Z9/9Nio69JU6udMZj/goap6CqCqHqqq+7e1s6paBXwQeFe3fn1VPdntvhnYf7jjkpyTZDDJ4Pr167f19NK08eKD9hh1XZpK/QyZ64B5Se5M8okkiyehz1uBQ4fZ/g7gq8MdUFWXVtVAVQ3MnTt3EkqQ+uukMw7jJQfvyU47hZccvCcnnXFYv0vSDqxvl8uq6okkxwDHAycCVyZ53wS7zRYbkrcBA8BkhJg07e0xd1dO+U9H97sMCejzu8uq6llgObA8ye3AmRPschHwr3c5k7wWuABYvOmynCRp6vTtclmSQ5IcPGTTQmDtBPo7EvgA8Dfd+iLgk8AbqurBidQqSdo2/ZzJ7AZcnGRPYCNwN3AOsBT4cpJnunY3VdWbRujj+CQrgdnAg8C7u3eWAXy8O8eXkgDcW1VvaDMUSdJw+nlPZgVw3DC7Thjj8cuBEd82U1Wv3abCJEmTxsfKSJKamfaPlUnyOuBjm22+p6pO6Uc9kqSxm/YhU1XXAtf2uw5J0vh5uUyS1IwhI0lqxpCRJDVjyEiSmjFkJEnNGDKSpGYMGUlSM4aMJKkZQ0aS1IwhI0lqxpCRJDVjyEiSmjFkJEnNGDKSpGYMGUlSM4aMJKkZQ0aS1IwhI0lqxpCRJDVjyEiSmjFkJEnNGDKSpGYMGUlSM4aMJKkZQ0aS1IwhI0lqxpCRJDVjyEiSmjFkJEnNGDKSpGYMGUlSM4aMJKkZQ0aS1IwhI0lqxpCRJDVjyEiSmpnykEnyxGbrZyW5pFuen2RZktuSLE+y/yj9LEiyIcnKJGuS3JLkzCH7k+Svk9zd9Xd0u1FJO5ZHH/hnrvyz9/Hffm8JV/7Z+3j0gX/ud0mapqbbTOYi4PKqOhL4IPCRrbT/UVUtqqrDgNOBP05ydrfv3wEHd69zgL9tVLO0w7n2b/+S+9Z8j+eefZb71nyPa//2L/tdkqap6RYyhwPLuuXrgSVjPbCqfgy8B3h3t2kJvcCqqroZ2DPJfpNZrLSjuv/ONaOuS5v0I2R2TbJq04vejGWT1cCp3fIpwO5J9h5H37cCh3bLLwXWDdl3X7ftlyQ5J8lgksH169eP41TSjuslLzts1HVpk36EzIaqWrjpBVw4ZN95wOIkK4HFwE+BjePoOyMsb1JbbKi6tKoGqmpg7ty54ziVtON63Tv/I/sfdgQ77bwz+x92BK9753/sd0mapmb1u4Chqup+4HcAkuwGnFpVj42ji0XApnn7fcC8Ifv2B+6fjDqlHd2e+76YN//ZR/tdhmaAaXVPJsk+STbVdD7wmXEcu4DeGwcu7jZdA5zRvcvsN4DHqupnk1iuJGkrptVMBjgB+EiSAm4Azt1K+4O6S2svAB4HLq6qy7p9XwF+G7gbeBI4e/guJEmtpGqL2xQ7rIGBgRocHOx3GZI0oyRZUVUDw+2bVpfLJEnbl+l2uWwLSV4OXLHZ5qeq6th+1CNJGrtpHzJVdTuwsN91SJLGz8tlkqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpmVn9LkCS+mXjwxt4ZOmdPL32cXaZvztzTnsZs/betd9lbVeaz2SSXJDkjiS3JVmV5Ngky5MMDNP2FUluSPLDJD9I8ukks0fo96wk65OsTHJXkmuTHDdk/8e7Pm5L8g9J9mw5TkkzzyNL7+Tpe34BzxVP3/MLHll6Z79L2u40DZkkrwROBo6uqiOB1wLrRmi7L/Al4L1VdQhwGPA1YPdRTnFlVS2qqoOBjwJXJTms2/d14IjuvHcC50/GmCRtP55e+/io65q41jOZ/YCHquopgKp6qKruH6HtucDnquqmrm1V1dKqemAsJ6qq64FLgXO69euqamO3+2Zg/wmMQ9J2aJf5u4+6rolrHTLXAfOS3JnkE0kWj9L2CGDFBM93K3DoMNt/H/jqcAckOSfJYJLB9evXT/D0kmaSOae9jF0OfCHsFHY58IXMOe1l/S5pu9P0xn9VPZHkGOB44ETgyiTva3jKbLEhuQDYCHx+hBovpTcDYmBgoBrWJmmambX3rrzoPxzV7zK2a83fXVZVzwLLgeVJbgfOHKHpHcAxwNUTON0iYM2mlSRn0rsn9JqqMkAkaYq1vvF/SJKDh2xaCKwdofklwJlJjh1y/NuSvHiM51pM737Mp7r11wPvBd5QVU9uS/2SpIlpPZPZDbi4e/vwRuBuekGwFPhykme6djdV1ZuSnA5clORFwHPADcBVo/T/5iS/CcwG7gFOrapNM5lLgOcDX08CcHNV/cHkDk+SNJp4Fen/GBgYqMHBwX6XIUkzSpIVVbXF7z6Cj5WRJDU07R8rk+Rs4I8223xjVZ3bj3okSWM37UOmqi4DLut3HZKk8fNymSSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkpoxZCRJzRgykqRmDBlJUjOGjCSpGUNGktSMISNJasaQkSQ1Y8hIkprpa8gkuSDJHUluS7IqybFJlicZGMOxJyR5LMnKJD9MckOSk4fsf36SK5PcneQ7SRa0HIskaUuz+nXiJK8ETgaOrqqnkuwD7DLObr5VVSd3/S0E/jHJhqpaBrwD+HlV/VqS04GPAW+exCFIUt888sgjXH311axbt4558+axZMkS5syZ0++yttDPmcx+wENV9RRAVT1UVfdva2dVtQr4IPCubtMS4HPd8lLgNUmy+XFJzkkymGRw/fr123p6SZpSV199NWvXruW5555j7dq1XH311f0uaVj9DJnrgHlJ7kzyiSSLJ6HPW4FDu+WXAusAqmoj8Biw9+YHVNWlVTVQVQNz586dhBIkqb1169aNuj5d9C1kquoJ4BjgHGA9cGWSsybYbUZY/tfTTrB/SZoW5s2bN+r6dNHXG/9V9WxVLa+q/0zvMtepE+xyEbCmW74PmAeQZBawB/DIBPuXpGlhyZIlzJ8/n5122on58+ezZMmSfpc0rH7e+D8EeK6q7uo2LQTWAkdsY39HAh8A/q9u0zXAmcBNwGnAN6rKmYyk7cKcOXM4++yz+13GVvUtZIDdgIuT7AlsBO6md+lsKfDlJM907W6qqjeN0MfxSVYCs4EHgXd37ywD+DvgiiR305vBnN5oHJKkEfQtZKpqBXDcMLtOGOPxy+ldAhtp//8GRgonSdIU8Df+JUnN9PNy2ZgkeR29X6Qc6p6qOqUf9UiSxm7ah0xVXQtc2+86JEnj5+UySVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWrGkJEkNWPISJKaMWQkSc0YMpKkZgwZSVIzhowkqRlDRpLUjCEjSWqmecgkuSDJHUluS7IqybFJlicZGKbtK5LckOSHSX6Q5NNJZo/Q71lJ1idZmeSuJNcmOW7I/jlJvt7t+3qSvVqOU5K0paYhk+SVwMnA0VV1JPBaYN0IbfcFvgS8t6oOAQ4DvgbsPsoprqyqRVV1MPBR4Kokh3X73gcs6/Yt69YlSVOo9UxmP+ChqnoKoKoeqqr7R2h7LvC5qrqpa1tVtbSqHhjLiarqeuBS4Jxu0xLgc93y54A3buMYJEnbqHXIXAfMS3Jnkk8kWTxK2yOAFRM8363Aod3yvlX1M4Du64uGOyDJOUkGkwyuX79+gqeXJA3VNGSq6gngGHqzi/XAlUnOanjKjPeAqrq0qgaqamDu3LktapKkHVbzG/9V9WxVLa+q/wy8Czh1hKZ30AukiVgErOmWH0iyH0D39cEJ9i1J24UNG+5lxa1v4RvXH8KKW9/Chg33NjtX6xv/hyQ5eMimhcDaEZpfApyZ5Nghx78tyYvHeK7F9GZMn+o2XQOc2S2fCVw9ntolaXv1/TXv5dFHb6FqI48+egvfX/PeZuea1aznnt2Ai5PsCWwE7qYXBEuBLyd5pmt3U1W9KcnpwEVJXgQ8B9wAXDVK/29O8pvAbOAe4NSq2jST+SjwxSTvAO4F3jTJY5OkGemxx24ddX0yNQ2ZqloBHDfMrhNGaH8TcPwY+/4s8NlR9j8MvGYsfUnSjmSPPY7m0Udv+aX1VvyNf0nawRx+2MfYc89XkMxizz1fweGHfazZuVpfLpuwJGcDf7TZ5hur6tx+1CNJM92uux7AMUf/jyk517QPmaq6DLis33VIksbPy2WSpGYMGUlSM4aMJKkZQ0aS1IwhI0lqxpCRJDVjyEiSmjFkJEnNGDKSpGYMGUlSM4aMJKkZQ0aS1IwhI0lqxpCRJDVjyEiSmklV9buGaSPJemBtH0vYB3ioj+efqJlePziG6WKmj2Gm1w/jG8P8qpo73A5DZhpJMlhVA/2uY1vN9PrBMUwXM30MM71+mLwxeLlMktSMISNJasaQmV4u7XcBEzTT6wfHMF3M9DHM9PphksbgPRlJUjPOZCRJzRgyUyDJ65P8MMndSd43zP7nJ7my2/+dJAuG7DsyyU1J7khye5IXTGXtQ+rYpjEkeV6Sz3W1r0ly/lTXPqTGrY3h1UluTbIxyWmb7TszyV3d68ypq/qXatim+pMsHPI9dFuSN09t5b9U4zb/HXT7X5jkp0kumZqKtzTB76MDklzX/Vv4/tB/61NpgmP4i+57aU2Sv06SUU9WVb4avoCdgR8BvwrsAqwGDt+szR8C/71bPh24slueBdwGHNWt7w3sPMPG8HvAF7rl2cBPgAXTdAwLgCOBy4HThmyfA/y4+7pXt7zXDKr/ZcDB3fJLgJ8Be86kv4Mh+/8K+H+BS6a6/skYA7Ac+K1ueTdg9kwaA3AccGPXx87ATcAJo53PmUx7rwDurqofV9XTwBeAJZu1WQJ8rlteCrym++ng3wK3VdVqgKp6uKqenaK6h5rIGAr4lSSzgF2Bp4FfTE3Zv2SrY6iqn1TVbcBzmx37OuDrVfVIVf0c+Drw+qkoeohtrr+q7qyqu7rl+4EHgWF/ca6xifwdkOQYYF/guqkodgTbPIYkhwOzqurrXbsnqurJKap7qIn8PRTwAnrh9HzgecADo53MkGnvpcC6Iev3dduGbVNVG4HH6M1aXgZUkmu7qev/MwX1DmciY1gK/Au9n57vBS6qqkdaFzyMsYyhxbGTZVJqSPIKev9B/GiS6hqPbR5Dkp2A/wL8SYO6xmMifw8vAx5NclWSlUk+nmTnSa9w67Z5DFV1E3A9vX/PPwOurao1ox1jyLQ33PXKzd/SN1KbWcBvAm/tvp6S5DWTW96YTGQMrwCepXeZ5kDgPyX51cktb0zGMoYWx06WCdeQZD/gCuDsqtpipjAFJjKGPwS+UlXrttqyrYmMYRZwPHAe8Ov0LledNTlljcs2jyHJrwGHAfvTC6aTkrx6tGMMmfbuA+YNWd8fuH+kNt1lpT2AR7rt36yqh7pp9VeAo5tXvKWJjOH3gK9V1TNV9SC967n9eNzGWMbQ4tjJMqEakrwQ+DLwp1V18yTXNlYTGcMrgXcl+QlwEfH+LnMAAARkSURBVHBGko9ObnljMtHvo5XdZaqNwD8yff89j+QU4ObuUt8TwFeB3xjtAEOmve8CByc5MMku9G6KX7NZm2uATe9YOg34RvXusl0LHJlkdvcf92Lg+1NU91ATGcO99H7aSZJfofcN+YMpqnuosYxhJNcC/zbJXkn2onev7NpGdY5km+vv2v8DcHlVfalhjVuzzWOoqrdW1QFVtYDeTODyqtriXVFTYCLfR98F9kqy6X7YSUzff88juRdYnGRWkufR+z9p1MtlU/7ujB3xBfw2cCe96+AXdNs+CLyhW34B8CXgbuAW4FeHHPs24A7ge8BfzLQx0HsHzZe6MXwf+JNpPIZfp/dT3r8ADwN3DDn297ux3U3vctOMqb/7HnoGWDXktXAmjWGzPs6iT+8um4Tvo9+i947R24HPArvMpDHQe0fZJ+kFy/eB/7q1c/kb/5KkZrxcJklqxpCRJDVjyEiSmjFkJEnNGDKSpGYMGUlSM4aMNIWSfHsbjvlJ91EJq7vHxL94vG277ft0y88mWZXke0n+Z5I9t31E0ugMGWkCuicxjFlVHbeNpzqxqo4CBoH3T7DthqpaWFVH0Hv0z7nbWJO0VYaMdnhJFiT5QXofrnZbkqXdo3yOSfLNJCu6J2Hv17VfnuTPk3wT+KMk+yb5h272sDrJiEGS5Inu635Jbhgyozh+jOXeAPzaJLa9ie4JvEl2S7Kse+L37Un+9fHvSc7o/mxWJ7mi2zbmcWvHNa6fwqTt2CHAO6rqxiSfoffT/SnAkqpan96nSX6Y3uNloPehX4sBklxJ70Gmp3SPbt9tDOf7PXqPSf9wd8zsMdZ5Mr1Hkky4bXfe1wB/123638ApVfWL7tLazUmuAQ4HLgBeVVUPJZnTtf9rxj9u7WAMGalnXVXd2C3/Pb3LTEcAX+999ho70/v8jE2uHLJ8EnAGQPU+VO6xMZzvu8BnuocM/mNVrdpK++uTPEvvuVd/OsG2uyZZRe/TD1fQ+xA26D0C/s+7R7c/R2+Gsy+98S2tqocA6v98HtC2jFs7GENG6tn8IX6P03so4CtHaP8vEzpZ1Q3df+b/Hrgiycer6vJRDjlx03/yY7C1thuqamGSPYB/ojdr+2t6n1s0Fzimqp7pHqv/Anrh40MOtU28JyP1HJBkU6C8BbgZmLtpW5LnJfk3Ixy7DHhn127n7rNbRpVkPvBgVX2K3uWqKf9ckap6DHg3cF43o9qjq+mZJCcC87umy4DfTbJ3V/ucIdvHNW7teAwZqWcNcGaS24A5wMX0PhfnY0lW03s8/kg3tv8IODHJ7fQuP40URkOdAKxKshI4FfiriZW/bapqJbCa3meKfB4YSDJIb1bzg67NHfTuR32z+7P4r93h2zJu7WB81L92eEkWAP/UvaVX0iRyJiNJasaZjDTJunsXy4bZ9ZqqeniU474DPH+zzW+vqi3ehjyetlI/GTKSpGa8XCZJasaQkSQ1Y8hIkpoxZCRJzRgykqRm/n9ks40tqaJ1rAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x576 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(6, 8))\n",
    "sns.stripplot(y = numbers_P_PIRacc.sort_values('perc_is_P_PIRacc').index, x= numbers_P_PIRacc.sort_values('perc_is_P_PIRacc').perc_is_P_PIRacc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fc7dcfa7550>"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZ0AAAHhCAYAAAC1P2MYAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3df5jedX3n++cLUhQKCwYiooQEKYWwCAlMQbAW/HHqj8NlpMAlbFWg9squhx6rHqpSVnZXt4tWrlMtiNtoBXE9LcpSobUKHiSLqwid/OKHKb+MMQhCAEFcKBry3j/ub2AMM5OZzMznniHPx3Xd13y/n+/n+/kBA698Pvc3952qQpKkFnbo9wAkSdsPQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNzOr3AKaTvfbaq+bPn9/vYUjSjLJ8+fKHqmrOWOoaOkPMnz+fwcHBfg9DkmaUJOvGWtftNUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM35z6AT96OEnOPuK1axY91OOmPciLjj5cPbbc5d+D0uSpiVXOhN09hWruXntI2zcVNy89hHOvmJ1v4ckSdPWlIVOkp9vcX5Gkou643lJrktyS5JlSfYdpZ35SZ5MsjLJmiQ3Jzl9yPXf79q5Jcl3kxw+5Nr7ktye5LYkf5PkhZM9zxXrfjrquSTpWf1a6VwAXFZVhwEfAc7fSv17qmpRVS0ATgXel+TM7tpa4LiurY8CSwGSvAx4DzBQVYcCO3b3Tqoj5r1o1HNJ0rP6FTqHANd1x9cDi8d6Y1X9AHg/vUChqr5bVZuXF98Dhq6aZgE7J5kF7ALcN8FxP8cFJx/OUfvPZtYO4aj9Z3PByYdv/SZJ2k5N5YMEOydZNeR8NnB1d7waOAn4FHAisFuSPavq4TG2vQI4eJjydwFfB6iqHye5APgR8CRwbVVdu+UNSZYASwD222+/MXb/rP323IUv/9tjxn2fJG2PpnKl82RVLdz8As4bcu1s4LgkK4HjgB8DG8fRdp5TkLyGXuh8sDt/Eb0V1P7AS4FfT/L2Le+rqqVVNVBVA3PmzBnHECRJ49WX7bWquq+qfq+qFgHndmWPjaOJRcCazSdJDgM+Byweslp6PbC2qjZU1S+BK4FjJ2UCkqRt0pfQSbJXks19nwN8fhz3zqf3IMKF3fl+9ALlHVV155CqPwJemWSXJAFex5CgkiS116+/HHo8cH6SAm4AztpK/QO6rbgXAo8DF1bVJd2184A9gYt72cLGbrvspiRX0Hv/ZyOwku7JNklSf6Sq+j2GaWNgYKAGBwf7PQxJmlGSLK+qgbHU9RMJJEnNTJvPXkvyCuCLWxQ/VVVH92M8kqTJN21Cp6puBRb2exySpKnj9pokqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JkKj6yFS94MH9mz9/ORtf0ekSRNC4bOVLjqLFj3Hdi0sffzqrP6PSJJmhamLHSS/HyL8zOSXNQdz0tyXZJbkixLsu8o7cxP8mSSlUnWJLk5yelDrh+c5MYkTyU5e4t735jkjiR3J/nQZM9xROtvGv1ckrZT/VrpXABcVlWHAR8Bzt9K/XuqalFVLQBOBd6X5Mzu2iPAe7o2n5FkR+DTwJuAQ4DTkhwyiXMY2dyjRz+XpO1Uv0LnEOC67vh6YPFYb6yqHwDvpxc0VNWDVfVPwC+3qHoUcHdV/aCqfgH87XD9JFmSZDDJ4IYNG8Y/k+Es/jTMexXsMKv3c/GnJ6ddSZrhZk1h2zsnWTXkfDZwdXe8GjgJ+BRwIrBbkj2r6uExtr0COHgrdV4GrB9yfi/wnCVHVS0FlgIMDAzUGPsf3ez94cx/nJSmJOn5ZCpXOk9W1cLNL+C8IdfOBo5LshI4DvgxsHEcbWcb60xOqEiStslUrnRGVFX3Ab8HkGRX4KSqemwcTSwC1mylzr3A3CHn+wL3jWeckqTJ1Zf3dJLslWRz3+cAnx/HvfPpPTRw4Vaq/hNwYJL9k+xE7wGEq7dyjyRpCvVlpQMcD5yfpIAbgK39RZYDuq24FwKPAxdW1SUASV4CDAL/CtiU5L3AIVX1syR/BFwD7Ah8vqpun5LZSJLGJFW+zbHZwMBADQ4O9nsYkjSjJFleVQNjqesnEkiSmunX9tpzJHkF8MUtip+qKv9mpSQ9T0yb0KmqW4GF/R6HJGnquL0mSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoPI+tf3w9Z3zjDBZdtogzvnEG6x9f3+8hSdrOjSl0kpyYpJIcvJV6fzo5wxq1j7cmOWSq+3k++PB3PszyB5azsTay/IHlfPg7H+73kCRt58a60jkN+J/AqVupN2zopGeyVlVvBQydMVj94OpRzyWpta0GQZJdgVcB76ILnST7JLkhyaoktyV5dZKPATt3ZV9KMj/JmiQXAyuAuUlOS3Jrd8/Hh/TxxiQrkqxOcl2SHZLclWROd32HJHcn+R3gLcAnun4O6F7fSLI8ybc3r8aSnNL1szrJDZP9D24mOPzFh496LkmtjWX18VbgG1V1J/BIkiOAfwNcU1ULgcOBVVX1IeDJqlpYVb/f3XsQcFlVLQJ+CXwceC2wEPitbqtsDvBZ4KSqOhw4pao2Af8N2NzO64HVVXUDcDXwJ10/9wBLgf+7qo4EzgYu7u45D3hD1+ZbtvGfz4z20Vd9lCP3PpJZmcWRex/JR1/10X4PSdJ2btYY6pwGfLI7/tvu/O+Bzyf5NeCrVbVqhHvXVdX3uuPfApZV1QaAJF8Cfgd4GrihqtYCVNUjXf3PA1d1ff8BcMmWjXersGOBryTZXPyC7ud3gEuTfBm4cqTJJVkCLAHYb7/9Rqo2I83dbS6XvvHSfg9Dkp4xaugk2ZPeyuTQJAXsCBTwAXqB8X8CX0zyiaq6bJgm/tfQ5kbqpmvzV1TV+iQPJHktcDTPrnqG2gF4tFtxbXn/v0tydDfGVUkWVtXDw9RbSm+1xMDAwHPGIUmaPFvbXjuZ3vbYvKqaX1VzgbX0AufBqvos8NfAEV39X3arn+HcBByXZK8kO9JbMf0P4MaufH+AJLOH3PM5ettsX66qp7uyx4HdAKrqZ8DaJKd09ybJ4d3xAVV1U1WdBzwEzB3LPxBJ0tTZWuicBvzdFmX/HbiU3uphJXAS8Knu2lLglm7r7FdU1f3AOcD1wGpgRVVd1W23LQGuTLIauHzIbVcDu/KrW2t/C/xJkpVJDqC3AnpXd+/twOKu3ic2P7QA3ND1KUnqo1RN3x2lJAPAX1TVq1v0NzAwUIODgy26kqTnjSTLq2pgLHXH8iBBXyT5EPBuhn8vR5I0A03bj8Gpqo917yX9z36PRZI0OaZt6EiSnn8MHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1Iyhs537xfr1rHv7O1hz6CtY9/Z38Iv16/s9JEnPY4bOdu7+c/6UJwYHYeNGnhgc5P5z/rTfQ5L0PNY8dJL8fIvzM5Jc1B3PS3JdkluSLEuy7yjtzE/yZJKVSdYkuTnJ6UOuJ8lfJrm7a++IqZvVzPXEqlWjnkvSZJpuK50LgMuq6jDgI8D5W6l/T1UtqqoFwKnA+5Kc2V17E3Bg91oCfGaKxjyj7bJw4ajnkjSZplvoHAJc1x1fDywe641V9QPg/cB7uqLF9AKsqup7wB5J9pnMwT4f7HP+f2GXgQGYNYtdBgbY5/z/0u8hSXoem9WHPndOMnQPZzZwdXe8GjgJ+BRwIrBbkj2r6uExtr0COLg7fhkw9F3xe7uy+4fekGQJvZUQ++233zim8fyw09y5zPtvX+z3MCRtJ/qx0nmyqhZufgHnDbl2NnBckpXAccCPgY3jaDsjHG9WzymoWlpVA1U1MGfOnHF0JUkar36sdEZUVfcBvweQZFfgpKp6bBxNLALWdMf3AnOHXNsXuG8yxilJ2jbT6j2dJHsl2Tymc4DPj+Pe+fQeRLiwK7oaeGf3FNsrgceq6v4RbpckNTCtVjrA8cD5SQq4AThrK/UP6LbiXgg8DlxYVZd01/4ReDNwN/AEcObwTUiSWknVc97m2G4NDAzU4OBgv4chSTNKkuVVNTCWutNqe02S9Pw23bbXniPJK4Atn+l9qqqO7sd4JEnbbtqHTlXdCvjX5CXpecDtNUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDXT19BJcm6S25PckmRVkqOTLEsyMIZ7j0/yWJKVSe5IckOSE4Zcf0GSy5PcneSmJPOnci6SpK2b1a+OkxwDnAAcUVVPJdkL2GmczXy7qk7o2lsIfDXJk1V1HfAu4KdV9RtJTgU+DrxtEqewXXtsw5N867I1/OSex3jJAbvz2ncuYPc5O/d7WJKmuX6udPYBHqqqpwCq6qGqum9bG6uqVcBHgD/qihYDX+iOrwBelyRb3pdkSZLBJIMbNmzY1u63O9+6bA333fUomzYV9931KN+6bE2/hyRpBuhn6FwLzE1yZ5KLkxw3CW2uAA7ujl8GrAeoqo3AY8CeW95QVUuraqCqBubMmTMJQ9g+/OSex0Y9l6Th9C10qurnwJHAEmADcHmSMybYbEY4fqbbCbavzksO2H3Uc0kaTl8fJKiqp6tqWVX9B3rbYidNsMlFwOZ9nnuBuQBJZgG7A49MsH11XvvOBbz0wD3YYYfw0gP34LXvXNDvIUmaAfr5IMFBwKaquqsrWgisAw7dxvYOAz4M/GFXdDVwOnAjcDLwrapypTNJdp+zMyf+P0f0exiSZpi+hQ6wK3Bhkj2AjcDd9LbargC+luSXXb0bq+qUEdp4dZKVwC7Ag8B7uifXAP4a+GKSu+mtcE6donlIksaob6FTVcuBY4e5dPwY719Gb8tspOv/AowUVpKkPvATCSRJzfRze21MkryB3l/sHGptVZ3Yj/FIkrbdtA+dqroGuKbf45AkTZzba5KkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUzJSHTpJzk9ye5JYkq5IcnWRZkoFh6h6V5IYkdyT55ySfS7LLCO2ekWRDkpVJ7kpyTZJjh1yfneSb3bVvJnnRVM5TkrR1Uxo6SY4BTgCOqKrDgNcD60eouzfwFeCDVXUQsAD4BrDbKF1cXlWLqupA4GPAlUkWdNc+BFzXXbuuO5ck9dFUr3T2AR6qqqcAquqhqrpvhLpnAV+oqhu7ulVVV1TVA2PpqKquB5YCS7qixcAXuuMvAG/dxjlIkibJVIfOtcDcJHcmuTjJcaPUPRRYPsH+VgAHd8d7V9X9AN3PFw93Q5IlSQaTDG7YsGGC3UuSRjOloVNVPweOpLf62ABcnuSMKewy472hqpZW1UBVDcyZM2cqxiRJ6sya6g6q6mlgGbAsya3A6SNUvZ1eQF01ge4WAWu64weS7FNV9yfZB3hwAu1KGsGjD/yEaz7zSe67cw0v/c0FvOHd72WPvV/S72FpmprqBwkOSnLgkKKFwLoRql8EnJ7k6CH3vz3JmH57u627JcBnu6KreTbgTmdiYSZpBNd85pPcu+Y2Nj39NPeuuY1rPvPJfg9J09hUr3R2BS5MsgewEbibXjBcAXwtyS+7ejdW1SlJTgUuSPJiYBNwA3DlKO2/LclvA7sAa4GTqmrzSudjwJeTvAv4EXDKJM9NEnDfnWtGPZeGmtLQqarlwLHDXDp+hPo3Aq8eY9uXApeOcv1h4HVjaUvStnvpby7g3jW3/cq5NBI/kUDShLzh3e9l3wWHssOOO7LvgkN5w7vf2+8haRqb8gcJJirJmcAfb1H8nao6qx/jkfSr9tj7JbztP36s38PQDDHtQ6eqLgEu6fc4JEkT5/aaJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqxtCRJDXT19BJcm6S25PckmRVkqOTLEsyMIZ7j0/yWJKVSe5IckOSE4Zcf3+S73dtX5dk3tTORpK0NbP61XGSY4ATgCOq6qkkewE7jbOZb1fVCV17C4GvJnmyqq4DVgIDVfVEkncDfw68bRKnIElTYuPDT/LIFXfyi3WPs9O83Zh98m8ya8+d+z2sSdHPlc4+wENV9RRAVT1UVfdta2NVtQr4CPBH3fn1VfVEd/l7wL4THK8kNfHIFXfyi7U/g03FL9b+jEeuuLPfQ5o0/Qyda4G5Se5McnGS4yahzRXAwcOUvwv4+nA3JFmSZDDJ4IYNGyZhCJI0Mb9Y9/io5zNZ30Knqn4OHAksATYAlyc5Y4LN5jkFyduBAeATI4xjaVUNVNXAnDlzJti9JE3cTvN2G/V8JuvbezoAVfU0sAxYluRW4PQJNrkIWLP5JMnrgXOB4zZv40nSdDf75N98zns6zxf9fJDgIGBTVd3VFS0E1gGHbmN7hwEfBv6wO18E/BXwxqp6cOIjlqQ2Zu25My/+t4f3exhTop8rnV2BC5PsAWwE7qa31XYF8LUkv+zq3VhVp4zQxquTrAR2AR4E3tM9uQa97bRdga8kAfhRVb1laqYiSRqLvoVOVS0Hjh3m0vFjvH8ZsPso11+/TQOTJE0ZP5FAktRMXx8kGIskbwA+vkXx2qo6sR/jkSRtu2kfOlV1DXBNv8chSZo4t9ckSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUzJSHTpJzk9ye5JYkq5IcnWRZkoFh6h6V5IYkdyT55ySfS7LLCO2ekWRDkpVJ7kpyTZJjh1z/RNfGLUn+LskeUzlPSdLWTWnoJDkGOAE4oqoOA14PrB+h7t7AV4APVtVBwALgG8Buo3RxeVUtqqoDgY8BVyZZ0F37JnBo1++dwDmTMSdJ0rab6pXOPsBDVfUUQFU9VFX3jVD3LOALVXVjV7eq6oqqemAsHVXV9cBSYEl3fm1Vbewufw/Yd7j7kixJMphkcMOGDWOemCRp/KY6dK4F5ia5M8nFSY4bpe6hwPIJ9rcCOHiY8j8Avj7cDVW1tKoGqmpgzpw5E+xekjSaKQ2dqvo5cCS91ccG4PIkZ0xhl3lOQXIusBH40hT2K0kag1lT3UFVPQ0sA5YluRU4fYSqt9MLqKsm0N0iYM3mkySn03tP6XVVVRNoV5L64pFHHuGqq65i/fr1zJ07l8WLFzN79ux+D2ubTfWDBAclOXBI0UJg3QjVLwJOT3L0kPvfnuQlY+zrOHorqs92528EPgi8paqe2JbxS1K/XXXVVaxbt45Nmzaxbt06rrpqIn8u77+pXunsClzYPa68EbibXjBcAXwtyS+7ejdW1SlJTgUuSPJiYBNwA3DlKO2/LclvA7sAa4GTqmrzSuci4AXAN5MAfK+q/t3kTk+Sptb69etHPZ9ppjR0qmo5cOwwl44fof6NwKvH2PalwKWjXP+NsbQjSdPZ3LlzWbdu3a+cz2R+IoEkTWOLFy9m3rx57LDDDsybN4/Fixf3e0gTMuUPEkxUkjOBP96i+DtVdVY/xiNJLc2ePZszzzyz38OYNNM+dKrqEuCSfo9DkjRxbq9JkpoxdCRJzRg6kqRmDB1JUjOGjiSpGUNHktSMoSNJasbQkSQ1Y+hIkpoxdCRJzRg6kqRmDB1JUjOGjiSpGUNHktSMoSNJasbQkSQ1Y+hIkpoxdCRJzRg6kqRmDB1JUjOGjiSpGUNHktSMoSNJasbQkSQ1Y+hIkpoxdCRJzRg6kqRmDB1JUjOGjiSpGUNHktSMoSNJasbQkSQ1Y+hIkpoxdCRJzRg6kqRmDB1JUjPNQyfJz7c4PyPJRd3xvCTXJbklybIk+47SzvwkTyZZmWRNkpuTnD7k+u937dyS5LtJDp+6WUmSxmJWvwewhQuAy6rqC0leC5wPvGOU+vdU1SKAJC8HrkyyQ1VdAqwFjquqnyZ5E7AUOHqKxy9JGsV02147BLiuO74eWDzWG6vqB8D7gfd059+tqp92l78HjLhqkiS10Y/Q2TnJqs0v4CNDrq0GTuqOTwR2S7LnONpeARw8TPm7gK8Pd0OSJUkGkwxu2LBhHF1JksarH6HzZFUt3PwCzhty7WzguCQrgeOAHwMbx9F2nlOQvIZe6HxwuBuqamlVDVTVwJw5c8bRlSRpvKbVezpVdR/wewBJdgVOqqrHxtHEImDN5pMkhwGfA95UVQ9P5lglSeM3rd7TSbJXks1jOgf4/DjunU/vQYQLu/P9gCuBd1TVnZM7UkmauZ588kcsX3Ea37r+IJavOI0nn/xRs76nVegAxwN3JLkT2Bv4s63UP2DzI9PAl4ELuyfXoLdttydwcff+0eBUDVqSZpLvr/kgjz56M1UbefTRm/n+mmHffZgSqapmnU13AwMDNThoNkl6fvvW9QdR9ezb5cksXvuaO7a5vSTLq2pgLHWn20pHkjTFdt/9iFHPp9K0D50krxj6iHX3uqnf45KkmeqQBR9njz2OIpnFHnscxSELPt6s72n19NpwqupWYGG/xyFJzxc777wfRx7xN33pe9qvdCRJzx+GjiSpGUNHktSMoSNJasbQkSQ1Y+hIkpoxdCRJzRg6kqRmDB1JUjOGjiSpGUNHktSMoSNJasbQkSQ1Y+hIkpoxdCRJzfh11UMk2QCsa9TdXsBDjfqaKs5henAO/TfTxw8Tm8O8qpozloqGTp8kGRzrd4pPV85henAO/TfTxw/t5uD2miSpGUNHktSModM/S/s9gEngHKYH59B/M3380GgOvqcjSWrGlY4kqRlDZ5ySfD7Jg0luG1I2O8k3k9zV/XxRV54kf5nk7iS3JDliyD2nd/XvSnL6kPIjk9za3fOXSTJaH9s4h7lJrk+yJsntSf54ps0jyQuT3JxkdTeH/9SV75/kpq79y5Ps1JW/oDu/u7s+f0hb53TldyR5w5DyN3Zldyf50JDyYfvYxnnsmGRlkn+YiePv2vth9+96VZLBrmwm/S7tkeSKJP/c/TdxzAwb/0HdP/vNr58lee+0nUNV+RrHC/gd4AjgtiFlfw58qDv+EPDx7vjNwNeBAK8EburKZwM/6H6+qDt+UXftZuCY7p6vA28arY9tnMM+wBHd8W7AncAhM2keXbu7dse/BtzUje3LwKld+X8F3t0d/1/Af+2OTwUu744PAVYDLwD2B+4Bduxe9wAvB3bq6hzS3TNsH9s4j/cD/x/wD6O1PV3H37XxQ2CvLcpm0u/SF4A/7I53AvaYSePfYi47Aj8B5k3XOfT9f+Iz8QXM51dD5w5gn+54H+CO7vivgNO2rAecBvzVkPK/6sr2Af55SPkz9UbqY5LmcxXwf8zUeQC7ACuAo+n95bZZXfkxwDXd8TXAMd3xrK5egHOAc4a0dU133zP3duXndK+M1Mc2jHtf4DrgtcA/jNb2dBz/kLZ/yHNDZ0b8LgH/ClhL9/72TBv/MPP5XeA703kObq9Njr2r6n6A7ueLu/KXAeuH1Lu3Kxut/N5hykfrY0K6bZpF9FYKM2oe3dbUKuBB4Jv0/mT/aFVtHKbfZ8baXX8M2HMb5rbnKH2M1yeBDwCbuvPR2p6O49+sgGuTLE+ypCubKb9LLwc2AJekt835uSS/PoPGv6VTgb/ZSvt9nYOhM7UyTFltQ/mUSLIr8N+B91bVz0arOkxZ3+dRVU9X1UJ6K4ajgAWj9DtZc5iUuSU5AXiwqpYPLR6l7Wk1/i28qqqOAN4EnJXkd0apOx3GO9Qsetvln6mqRcD/ordNNJLpNv5ndO/NvQX4ytaqDlPWbA6GzuR4IMk+AN3PB7vye4G5Q+rtC9QsyIMAAAT8SURBVNy3lfJ9hykfrY9tkuTX6AXOl6rqypk6D4CqehRYRm9/eo8ks4bp95mxdtd3Bx7Zhrk9NEof4/Eq4C1Jfgj8Lb0ttk/OoPE/o6ru634+CPwdvT8AzJTfpXuBe6vqpu78CnohNFPGP9SbgBVV9cBW2u/rHAydyXE1cHp3fDq990g2l7+ze1rklcBj3RL0GuB3k7yoe9rjd+ntq98PPJ7kld3TIe/coq3h+hi3ru2/BtZU1f87E+eRZE6SPbrjnYHXA2uA64GTR5jD5n5PBr5VvY3oq4FT03s6bH/gQHpvmv4TcGB6T3rtRG/b4urunpH6GLOqOqeq9q2q+V3b36qq358p498sya8n2W3zMb3fgduYIb9LVfUTYH2Sg7qi1wHfnynj38JpPLu1Nlr7/Z3DRN+42t5e3b/U+4Ff0vsTwLvo7ZNfB9zV/Zzd1Q3waXrvNdwKDAxp5w+Au7vXmUPKB+j9R3sPcBHP/gXeYfvYxjn8Nr3l8S3Aqu715pk0D+AwYGU3h9uA87ryl9P7n+7d9LYZXtCVv7A7v7u7/vIhbZ3bjfMOuqdyuvI303uy7x7g3CHlw/YxgX8fx/Ps02szavxdW6u71+2b+5lhv0sLgcHud+mr9J7cmjHj79raBXgY2H1I2bScg59IIElqxu01SVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXSkhpJ8JMnrx3nPf0zy4/Q+tv62JG8Zb92u/Ozu+NIka7s6q5O8bmKzksZu1tarSJosVXXeNt76F1V1QZIFwLeTvLiqNo217jB1/qSqrkjyGnpfU3zgNo5LGhdXOtruJZmf3pd3fTa9L4S7NsnOSZYlGejq7NV9ThpJzkjy1SR/360Y/ijJ+9P7lOLvJZk9Sl+XJjm5O/5Yku+n90VaF4xlrFW1BtgI7DVJdW9kyKdMJzkvyT91q6Sl3ceekOQ3kvz/3cpoRZIDuvIPpPflXquTfGwsc9D2zdCReg4EPl1V/xp4FDhpK/UPBf4NvQ+3/DPgiep9SvGN9D6balRdMJ0I/OuqOgz4z2MZZJKj6X0VwoZJqvtGeh/9stlFVfVbVXUosDNwQlf+JXr/fA4HjgXuT/Im4K3A0V35n49lDtq+ub0m9aytqlXd8XJ6X9Q3muur6nF6H4T4GPD3Xfmt9D4Xbmt+BvwL8LkkX6P3JW6jeV+StwOPA2+r0T+/6jl1uwXLUJ9I8uf0vv/klUPKX5PkA/Q+y2s2cHuSZcDLqurvAKrqXwC696YuqaonuvJHtjZpyZWO1PPUkOOn6f2BbCPP/jfywlHqbxpyvokx/GGuel+idhS9r5d4K/CNrdzyF1W1sKpeXVXfnoS6fwL8BvDv6X1dM0leCFwMnFxVrwA+S2/ew32fCl25H96ocTF0pJH9EDiyOz55lHrjlt4X6O1eVf8IvJfeJx031T2I8ClghyRv4Nlgfagb38ldvZ8B9yZ5azf2FyTZBbgW+IPuePOWoTQqQ0ca2QXAu5N8lzG8cT9OuwH/kOQW4H8A75vk9sek26b7z8AHqvdleJ+lt0X4VXrfybPZO4D3dOP9LvCSqvoGve9TGUzva8PPbjp4zUh+tYEkqRlXOpKkZnx6TZoCST4NvGqL4k9V1SWj3HMucMoWxV+pqj+bSF1pOnF7TZLUjNtrkqRmDB1JUjOGjiSpGUNHktSMoSNJauZ/A4u6vS3NHusqAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x576 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(6, 8))\n",
    "sns.stripplot(y = numbers_P_PIRacc.sort_values('num_is_P_PIRacc').index, x= numbers_P_PIRacc.sort_values('num_is_P_PIRacc').num_is_P_PIRacc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Astrocytes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:47: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H9_D0\n",
      "H9_D10\n",
      "H9_D2\n",
      "H9_D28\n",
      "SLC_D0\n",
      "SLC_D2\n",
      "SL_D0\n",
      "SL_D2\n",
      "CPU times: user 4min 35s, sys: 2.46 s, total: 4min 38s\n",
      "Wall time: 3min 30s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "P_PIR_gene_classification={}\n",
    "for loop_file in sorted(loop_files):\n",
    "    loop_filename = os.path.basename(loop_file)\n",
    "    tissue = loop_filename.split('.')[0]\n",
    "    print(tissue)\n",
    "    \n",
    "    # read loop/anchor files\n",
    "    loop_df = pd.read_csv(loop_file, index_col=0)\n",
    "    anchor_file = os.path.join(anchor_dir, tissue+'.anchors.csv')\n",
    "    anchor_df = pd.read_csv(anchor_file, index_col=0)\n",
    "    \n",
    "    # read anchor atac annotation\n",
    "    anchor_annon_file = os.path.join(anchor_annon_dir, tissue+'_annon.bed')\n",
    "    if not os.path.exists(anchor_annon_file):\n",
    "        print(tissue, 'not anchor annon file.. skipping')\n",
    "        continue\n",
    "    anchor_annon_df = pd.read_csv(anchor_annon_file, names=['chr', 'start', 'stop', 'name', 'atac_chr', 'atac_start', 'atac_stop'], sep='\\t')\n",
    "    anchor_to_count_dict = anchor_annon_df.groupby('name').name.count().to_dict()\n",
    "    anchors_acc = anchor_to_count_dict.keys()\n",
    "    \n",
    "    # read anchor tss annotation\n",
    "    promoter_anchor_file = os.path.join(promoter_annon_dir, 'promoter_'+tissue+'_annon.bed')\n",
    "    if not os.path.exists(promoter_anchor_file):\n",
    "        print(tissue, 'not promoter annon file.. skipping')\n",
    "        continue\n",
    "    promoter_anchor_df = pd.read_csv(promoter_anchor_file, names=['chr', 'start', 'stop', 'tss', 'anchor_chr', 'anchor_start', \n",
    "                                                             'anchor_stop', 'anchor', 'overlap'], sep='\\t')\n",
    "    promoter_anchors = promoter_anchor_df.anchor.unique()\n",
    "\n",
    "    \n",
    "    loop_df['source_P'] = loop_df.source.isin(promoter_anchors)\n",
    "    loop_df['target_P'] = loop_df.target.isin(promoter_anchors)\n",
    "    \n",
    "    loop_df['source_acc'] = loop_df.source.isin(anchors_acc)\n",
    "    loop_df['target_acc'] = loop_df.target.isin(anchors_acc)   \n",
    "\n",
    "    loop_df['source_acc_PIR'] = loop_df['source_acc'] & (~loop_df['source_P'])\n",
    "    loop_df['target_acc_PIR'] = loop_df['target_acc'] & (~loop_df['target_P'])\n",
    "    \n",
    "    loop_df['is_P_PIR'] = loop_df['source_P'] ^ loop_df['target_P']    \n",
    "    loop_df['is_P_PIRacc'] = loop_df['is_P_PIR'] & (loop_df['source_acc_PIR'] | loop_df['target_acc_PIR']) \n",
    "    loop_df['both_anchor_acc'] = loop_df['source_acc'] & loop_df['target_acc']\n",
    "    loop_df['is_Pacc_PIRacc'] = loop_df['is_P_PIR']  & loop_df['both_anchor_acc'] \n",
    "\n",
    "    loop_df_P_PIRacc = loop_df[loop_df.is_P_PIRacc]\n",
    "    loop_df_P_PIRacc['P_column'] = loop_df_P_PIRacc.apply(lambda x: x.source if x.source_P else x.target,axis=1)\n",
    "    loop_df_P_PIRacc['E_column'] = loop_df_P_PIRacc.apply(lambda x: x.target if x.source_P else x.source,axis=1)\n",
    "    # get the actual loop\n",
    "    loop_df_P_PIRacc = loop_df_P_PIRacc.merge(promoter_anchor_df[['anchor','tss']],how='left',left_on='P_column', right_on='anchor')\n",
    "\n",
    "    # get closest gene to PIRacc\n",
    "    loop_df_P_PIRacc = loop_df_P_PIRacc.merge(anchors_closest_gene_df[['anchor_name', 'gene', 'dist_to_TSS']],how='left',left_on='E_column', right_on='anchor_name')\n",
    "\n",
    "    # renaming colummns\n",
    "    loop_df_P_PIRacc = loop_df_P_PIRacc[['P_column', 'E_column', 'is_Pacc_PIRacc', 'tss', 'gene', 'dist_to_TSS']]\n",
    "    loop_df_P_PIRacc.columns = ['P_column', 'E_column', 'is_Pacc_PIRacc','tss_of_P', 'nearest_gene_to_PIR','dist_to_nearest_gene']\n",
    "\n",
    "    # get distal/nearest classification\n",
    "    loop_df_P_PIRacc['is_nearest_gene'] = loop_df_P_PIRacc.nearest_gene_to_PIR == loop_df_P_PIRacc.tss_of_P\n",
    "\n",
    "    # get numbers for P-PIRacc loops\n",
    "    P_PIRacc_types = loop_df_P_PIRacc.groupby(['E_column','nearest_gene_to_PIR']).agg({'tss_of_P':'::'.join,'P_column':'count'}).reset_index()\n",
    "    P_PIRacc_types['E_loops_to_nearest'] = P_PIRacc_types.apply(lambda x: x.nearest_gene_to_PIR in x.tss_of_P, axis=1)\n",
    "    P_PIRacc_types['type1_distal_only'] = (~P_PIRacc_types.E_loops_to_nearest)\n",
    "    P_PIRacc_types['type2_nearest_distal'] = P_PIRacc_types.E_loops_to_nearest & (P_PIRacc_types.P_column>1)\n",
    "    P_PIRacc_types['type3_nearest_only'] = P_PIRacc_types.E_loops_to_nearest & (P_PIRacc_types.P_column==1)\n",
    "    num_enhancers = len(P_PIRacc_types)\n",
    "    numE_type1_distal_only = len(P_PIRacc_types[P_PIRacc_types.type1_distal_only])\n",
    "    numE_type2_nearest_distal = len(P_PIRacc_types[P_PIRacc_types.type2_nearest_distal])\n",
    "    numE_type3_nearest_only = len(P_PIRacc_types[P_PIRacc_types.type3_nearest_only])    \n",
    "    num_target_genes = P_PIRacc_types.P_column.sum()\n",
    "    numP_type1_distal_only = P_PIRacc_types[P_PIRacc_types.type1_distal_only].P_column.sum()\n",
    "    numP_type2_nearest_distal = P_PIRacc_types[P_PIRacc_types.type2_nearest_distal].P_column.sum()\n",
    "    numP_type3_nearest_only = P_PIRacc_types[P_PIRacc_types.type3_nearest_only].P_column.sum()\n",
    "    \n",
    "\n",
    "    # get numbers for Pacc-PIRacc loops\n",
    "    loop_df_Pacc_PIRacc = loop_df_P_PIRacc[loop_df_P_PIRacc.is_Pacc_PIRacc]\n",
    "    Pacc_PIRacc_types = loop_df_Pacc_PIRacc.groupby(['E_column','nearest_gene_to_PIR']).agg({\n",
    "        'tss_of_P':'::'.join,'P_column':'count',}).reset_index()\n",
    "    Pacc_PIRacc_types['E_loops_to_nearest'] = Pacc_PIRacc_types.apply(lambda x: x.nearest_gene_to_PIR in x.tss_of_P, axis=1)\n",
    "    Pacc_PIRacc_types['type1_distal_only'] = (~Pacc_PIRacc_types.E_loops_to_nearest) \n",
    "    Pacc_PIRacc_types['type2_nearest_distal'] = Pacc_PIRacc_types.E_loops_to_nearest & (Pacc_PIRacc_types.P_column>1)\n",
    "    Pacc_PIRacc_types['type3_nearest_only'] = Pacc_PIRacc_types.E_loops_to_nearest & (Pacc_PIRacc_types.P_column==1)\n",
    "    num_enhancers_Pacc = len(Pacc_PIRacc_types)\n",
    "    numE_type1_distal_only_Pacc = len(Pacc_PIRacc_types[Pacc_PIRacc_types.type1_distal_only])\n",
    "    numE_type2_nearest_distal_Pacc = len(Pacc_PIRacc_types[Pacc_PIRacc_types.type2_nearest_distal])\n",
    "    numE_type3_nearest_only_Pacc = len(Pacc_PIRacc_types[Pacc_PIRacc_types.type3_nearest_only])\n",
    "    num_target_genes_Pacc = Pacc_PIRacc_types.P_column.sum()\n",
    "    numP_type1_distal_only_Pacc = Pacc_PIRacc_types[Pacc_PIRacc_types.type1_distal_only].P_column.sum()\n",
    "    numP_type2_nearest_distal_Pacc = Pacc_PIRacc_types[Pacc_PIRacc_types.type2_nearest_distal].P_column.sum()\n",
    "    numP_type3_nearest_only_Pacc = Pacc_PIRacc_types[Pacc_PIRacc_types.type3_nearest_only].P_column.sum()\n",
    "    \n",
    "    \n",
    "    P_PIR_gene_classification[tissue] = {'num_loops_P_PIRacc':len(loop_df_P_PIRacc),\n",
    "                                         'num_enhancers_in_P_PIRacc':num_enhancers,\n",
    "                                         'numE_type1_distal_only_in_P_PIRacc':numE_type1_distal_only,\n",
    "                                         'numE_type2_nearest_distal_in_P_PIRacc':numE_type2_nearest_distal,\n",
    "                                         'numE_type3_nearest_only_in_P_PIRacc':numE_type3_nearest_only,\n",
    "                                         'num_target_genes_in_P_PIRacc':num_target_genes,\n",
    "                                         'numP_type1_distal_only':numP_type1_distal_only,\n",
    "                                         'numP_type2_nearest_distal':numP_type2_nearest_distal,\n",
    "                                         'numP_type3_nearest_only':numP_type3_nearest_only,\n",
    "                                         \n",
    "                                         'num_loops_Pacc_PIRacc':len(loop_df_Pacc_PIRacc),\n",
    "                                         'num_enhancers_in_Pacc_PIRacc':num_enhancers_Pacc,\n",
    "                                         'numE_type1_distal_only_in_Pacc_PIRacc':numE_type1_distal_only_Pacc,\n",
    "                                         'numE_type2_nearest_distal_in_Pacc_PIRacc':numE_type2_nearest_distal_Pacc,\n",
    "                                         'numE_type3_nearest_only_in_Pacc_PIRacc':numE_type3_nearest_only_Pacc, \n",
    "                                         'num_target_genes_in_Pacc_PIRacc':num_target_genes_Pacc,\n",
    "                                         'numP_type1_distal_only_Pacc':numP_type1_distal_only_Pacc,\n",
    "                                         'numP_type2_nearest_distal_Pacc':numP_type2_nearest_distal_Pacc,\n",
    "                                         'numP_type3_nearest_only_Pacc':numP_type3_nearest_only_Pacc\n",
    "                                           \n",
    "    }   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_loops_P_PIRacc</th>\n",
       "      <th>num_enhancers_in_P_PIRacc</th>\n",
       "      <th>numE_type1_distal_only_in_P_PIRacc</th>\n",
       "      <th>numE_type2_nearest_distal_in_P_PIRacc</th>\n",
       "      <th>numE_type3_nearest_only_in_P_PIRacc</th>\n",
       "      <th>num_target_genes_in_P_PIRacc</th>\n",
       "      <th>numP_type1_distal_only</th>\n",
       "      <th>numP_type2_nearest_distal</th>\n",
       "      <th>numP_type3_nearest_only</th>\n",
       "      <th>num_loops_Pacc_PIRacc</th>\n",
       "      <th>num_enhancers_in_Pacc_PIRacc</th>\n",
       "      <th>numE_type1_distal_only_in_Pacc_PIRacc</th>\n",
       "      <th>numE_type2_nearest_distal_in_Pacc_PIRacc</th>\n",
       "      <th>numE_type3_nearest_only_in_Pacc_PIRacc</th>\n",
       "      <th>num_target_genes_in_Pacc_PIRacc</th>\n",
       "      <th>numP_type1_distal_only_Pacc</th>\n",
       "      <th>numP_type2_nearest_distal_Pacc</th>\n",
       "      <th>numP_type3_nearest_only_Pacc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>241235</td>\n",
       "      <td>15137</td>\n",
       "      <td>3188</td>\n",
       "      <td>11866</td>\n",
       "      <td>83</td>\n",
       "      <td>241235</td>\n",
       "      <td>40944</td>\n",
       "      <td>200208</td>\n",
       "      <td>83</td>\n",
       "      <td>37355</td>\n",
       "      <td>8855</td>\n",
       "      <td>7797</td>\n",
       "      <td>866</td>\n",
       "      <td>192</td>\n",
       "      <td>37355</td>\n",
       "      <td>30620</td>\n",
       "      <td>6543</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H9_D0</td>\n",
       "      <td>255247</td>\n",
       "      <td>30791</td>\n",
       "      <td>9647</td>\n",
       "      <td>19873</td>\n",
       "      <td>1271</td>\n",
       "      <td>255247</td>\n",
       "      <td>54057</td>\n",
       "      <td>199919</td>\n",
       "      <td>1271</td>\n",
       "      <td>55020</td>\n",
       "      <td>15984</td>\n",
       "      <td>12438</td>\n",
       "      <td>2543</td>\n",
       "      <td>1003</td>\n",
       "      <td>55020</td>\n",
       "      <td>38942</td>\n",
       "      <td>15075</td>\n",
       "      <td>1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H9_D10</td>\n",
       "      <td>133816</td>\n",
       "      <td>18146</td>\n",
       "      <td>6292</td>\n",
       "      <td>11013</td>\n",
       "      <td>841</td>\n",
       "      <td>133816</td>\n",
       "      <td>32460</td>\n",
       "      <td>100515</td>\n",
       "      <td>841</td>\n",
       "      <td>27618</td>\n",
       "      <td>8138</td>\n",
       "      <td>6732</td>\n",
       "      <td>1033</td>\n",
       "      <td>373</td>\n",
       "      <td>27618</td>\n",
       "      <td>20830</td>\n",
       "      <td>6415</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H9_D2</td>\n",
       "      <td>950434</td>\n",
       "      <td>45106</td>\n",
       "      <td>7522</td>\n",
       "      <td>37039</td>\n",
       "      <td>545</td>\n",
       "      <td>950434</td>\n",
       "      <td>138973</td>\n",
       "      <td>810916</td>\n",
       "      <td>545</td>\n",
       "      <td>143820</td>\n",
       "      <td>32851</td>\n",
       "      <td>27353</td>\n",
       "      <td>4215</td>\n",
       "      <td>1283</td>\n",
       "      <td>143820</td>\n",
       "      <td>113987</td>\n",
       "      <td>28550</td>\n",
       "      <td>1283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H9_D28</td>\n",
       "      <td>122050</td>\n",
       "      <td>14443</td>\n",
       "      <td>4378</td>\n",
       "      <td>9633</td>\n",
       "      <td>432</td>\n",
       "      <td>122050</td>\n",
       "      <td>26630</td>\n",
       "      <td>94988</td>\n",
       "      <td>432</td>\n",
       "      <td>14646</td>\n",
       "      <td>5156</td>\n",
       "      <td>4377</td>\n",
       "      <td>547</td>\n",
       "      <td>232</td>\n",
       "      <td>14646</td>\n",
       "      <td>11620</td>\n",
       "      <td>2794</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SLC_D0</td>\n",
       "      <td>672211</td>\n",
       "      <td>44888</td>\n",
       "      <td>11348</td>\n",
       "      <td>32717</td>\n",
       "      <td>823</td>\n",
       "      <td>672211</td>\n",
       "      <td>111826</td>\n",
       "      <td>559562</td>\n",
       "      <td>823</td>\n",
       "      <td>175823</td>\n",
       "      <td>32067</td>\n",
       "      <td>24833</td>\n",
       "      <td>6066</td>\n",
       "      <td>1168</td>\n",
       "      <td>175823</td>\n",
       "      <td>122225</td>\n",
       "      <td>52430</td>\n",
       "      <td>1168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SLC_D2</td>\n",
       "      <td>886043</td>\n",
       "      <td>57243</td>\n",
       "      <td>12612</td>\n",
       "      <td>43718</td>\n",
       "      <td>913</td>\n",
       "      <td>886043</td>\n",
       "      <td>145400</td>\n",
       "      <td>739730</td>\n",
       "      <td>913</td>\n",
       "      <td>188033</td>\n",
       "      <td>41116</td>\n",
       "      <td>32180</td>\n",
       "      <td>7095</td>\n",
       "      <td>1841</td>\n",
       "      <td>188033</td>\n",
       "      <td>135294</td>\n",
       "      <td>50898</td>\n",
       "      <td>1841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SL_D0</td>\n",
       "      <td>292572</td>\n",
       "      <td>38730</td>\n",
       "      <td>13888</td>\n",
       "      <td>23079</td>\n",
       "      <td>1763</td>\n",
       "      <td>292572</td>\n",
       "      <td>72384</td>\n",
       "      <td>218425</td>\n",
       "      <td>1763</td>\n",
       "      <td>74057</td>\n",
       "      <td>21843</td>\n",
       "      <td>16760</td>\n",
       "      <td>3648</td>\n",
       "      <td>1435</td>\n",
       "      <td>74057</td>\n",
       "      <td>51317</td>\n",
       "      <td>21305</td>\n",
       "      <td>1435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SL_D2</td>\n",
       "      <td>882508</td>\n",
       "      <td>49343</td>\n",
       "      <td>9478</td>\n",
       "      <td>39199</td>\n",
       "      <td>666</td>\n",
       "      <td>882508</td>\n",
       "      <td>141756</td>\n",
       "      <td>740086</td>\n",
       "      <td>666</td>\n",
       "      <td>155846</td>\n",
       "      <td>35160</td>\n",
       "      <td>28626</td>\n",
       "      <td>5103</td>\n",
       "      <td>1431</td>\n",
       "      <td>155846</td>\n",
       "      <td>118428</td>\n",
       "      <td>35987</td>\n",
       "      <td>1431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            num_loops_P_PIRacc  num_enhancers_in_P_PIRacc  \\\n",
       "Astrocytes              241235                      15137   \n",
       "H9_D0                   255247                      30791   \n",
       "H9_D10                  133816                      18146   \n",
       "H9_D2                   950434                      45106   \n",
       "H9_D28                  122050                      14443   \n",
       "SLC_D0                  672211                      44888   \n",
       "SLC_D2                  886043                      57243   \n",
       "SL_D0                   292572                      38730   \n",
       "SL_D2                   882508                      49343   \n",
       "\n",
       "            numE_type1_distal_only_in_P_PIRacc  \\\n",
       "Astrocytes                                3188   \n",
       "H9_D0                                     9647   \n",
       "H9_D10                                    6292   \n",
       "H9_D2                                     7522   \n",
       "H9_D28                                    4378   \n",
       "SLC_D0                                   11348   \n",
       "SLC_D2                                   12612   \n",
       "SL_D0                                    13888   \n",
       "SL_D2                                     9478   \n",
       "\n",
       "            numE_type2_nearest_distal_in_P_PIRacc  \\\n",
       "Astrocytes                                  11866   \n",
       "H9_D0                                       19873   \n",
       "H9_D10                                      11013   \n",
       "H9_D2                                       37039   \n",
       "H9_D28                                       9633   \n",
       "SLC_D0                                      32717   \n",
       "SLC_D2                                      43718   \n",
       "SL_D0                                       23079   \n",
       "SL_D2                                       39199   \n",
       "\n",
       "            numE_type3_nearest_only_in_P_PIRacc  num_target_genes_in_P_PIRacc  \\\n",
       "Astrocytes                                   83                        241235   \n",
       "H9_D0                                      1271                        255247   \n",
       "H9_D10                                      841                        133816   \n",
       "H9_D2                                       545                        950434   \n",
       "H9_D28                                      432                        122050   \n",
       "SLC_D0                                      823                        672211   \n",
       "SLC_D2                                      913                        886043   \n",
       "SL_D0                                      1763                        292572   \n",
       "SL_D2                                       666                        882508   \n",
       "\n",
       "            numP_type1_distal_only  numP_type2_nearest_distal  \\\n",
       "Astrocytes                   40944                     200208   \n",
       "H9_D0                        54057                     199919   \n",
       "H9_D10                       32460                     100515   \n",
       "H9_D2                       138973                     810916   \n",
       "H9_D28                       26630                      94988   \n",
       "SLC_D0                      111826                     559562   \n",
       "SLC_D2                      145400                     739730   \n",
       "SL_D0                        72384                     218425   \n",
       "SL_D2                       141756                     740086   \n",
       "\n",
       "            numP_type3_nearest_only  num_loops_Pacc_PIRacc  \\\n",
       "Astrocytes                       83                  37355   \n",
       "H9_D0                          1271                  55020   \n",
       "H9_D10                          841                  27618   \n",
       "H9_D2                           545                 143820   \n",
       "H9_D28                          432                  14646   \n",
       "SLC_D0                          823                 175823   \n",
       "SLC_D2                          913                 188033   \n",
       "SL_D0                          1763                  74057   \n",
       "SL_D2                           666                 155846   \n",
       "\n",
       "            num_enhancers_in_Pacc_PIRacc  \\\n",
       "Astrocytes                          8855   \n",
       "H9_D0                              15984   \n",
       "H9_D10                              8138   \n",
       "H9_D2                              32851   \n",
       "H9_D28                              5156   \n",
       "SLC_D0                             32067   \n",
       "SLC_D2                             41116   \n",
       "SL_D0                              21843   \n",
       "SL_D2                              35160   \n",
       "\n",
       "            numE_type1_distal_only_in_Pacc_PIRacc  \\\n",
       "Astrocytes                                   7797   \n",
       "H9_D0                                       12438   \n",
       "H9_D10                                       6732   \n",
       "H9_D2                                       27353   \n",
       "H9_D28                                       4377   \n",
       "SLC_D0                                      24833   \n",
       "SLC_D2                                      32180   \n",
       "SL_D0                                       16760   \n",
       "SL_D2                                       28626   \n",
       "\n",
       "            numE_type2_nearest_distal_in_Pacc_PIRacc  \\\n",
       "Astrocytes                                       866   \n",
       "H9_D0                                           2543   \n",
       "H9_D10                                          1033   \n",
       "H9_D2                                           4215   \n",
       "H9_D28                                           547   \n",
       "SLC_D0                                          6066   \n",
       "SLC_D2                                          7095   \n",
       "SL_D0                                           3648   \n",
       "SL_D2                                           5103   \n",
       "\n",
       "            numE_type3_nearest_only_in_Pacc_PIRacc  \\\n",
       "Astrocytes                                     192   \n",
       "H9_D0                                         1003   \n",
       "H9_D10                                         373   \n",
       "H9_D2                                         1283   \n",
       "H9_D28                                         232   \n",
       "SLC_D0                                        1168   \n",
       "SLC_D2                                        1841   \n",
       "SL_D0                                         1435   \n",
       "SL_D2                                         1431   \n",
       "\n",
       "            num_target_genes_in_Pacc_PIRacc  numP_type1_distal_only_Pacc  \\\n",
       "Astrocytes                            37355                        30620   \n",
       "H9_D0                                 55020                        38942   \n",
       "H9_D10                                27618                        20830   \n",
       "H9_D2                                143820                       113987   \n",
       "H9_D28                                14646                        11620   \n",
       "SLC_D0                               175823                       122225   \n",
       "SLC_D2                               188033                       135294   \n",
       "SL_D0                                 74057                        51317   \n",
       "SL_D2                                155846                       118428   \n",
       "\n",
       "            numP_type2_nearest_distal_Pacc  numP_type3_nearest_only_Pacc  \n",
       "Astrocytes                            6543                           192  \n",
       "H9_D0                                15075                          1003  \n",
       "H9_D10                                6415                           373  \n",
       "H9_D2                                28550                          1283  \n",
       "H9_D28                                2794                           232  \n",
       "SLC_D0                               52430                          1168  \n",
       "SLC_D2                               50898                          1841  \n",
       "SL_D0                                21305                          1435  \n",
       "SL_D2                                35987                          1431  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame.from_dict(P_PIR_gene_classification,orient='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "P_PIR_gene_classification_df = pd.DataFrame.from_dict(P_PIR_gene_classification,orient='index')\n",
    "P_PIR_gene_classification_df.to_csv(os.path.join(save_dir, 'P_PIR_gene_classification_df.csv'))\n",
    "P_PIR_gene_classification_normal_df = P_PIR_gene_classification_df[P_PIR_gene_classification_df.index.isin(normal_tissues)]\n",
    "P_PIR_gene_classification_normal_df.to_csv(os.path.join(save_dir, 'P_PIR_gene_classification_normal_df.csv'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. get snp - egene \n",
    "\n",
    "# 4A. create snp-egene table\n",
    "\n",
    "for each snp `name`=`rsid`_`disease_name` \n",
    "1. annotate to anchor regions with snp data\n",
    "2. annotate anchor regions with TSS data (already done just read in)\n",
    "3. merge with loops\n",
    "4. a SNP-egene pair where a SNP and egene are found on the same loop (regardless if found on same anchor point or different one)\n",
    "5. (NOT DONE BECAUSE 1615 combination of diseases, ###TODO make better tissue disease matcher ) add a potential tissue match\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(65019, 4)\n",
      "1575\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>78128047</td>\n",
       "      <td>78128048</td>\n",
       "      <td>rs881361__Alzheimer Dementia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>153652930</td>\n",
       "      <td>153652931</td>\n",
       "      <td>rs73281459__Schizophrenia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr5</td>\n",
       "      <td>153656738</td>\n",
       "      <td>153656739</td>\n",
       "      <td>rs73279671__Schizophrenia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr8</td>\n",
       "      <td>109260581</td>\n",
       "      <td>109260582</td>\n",
       "      <td>rs9969614__Dupuytren Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr10</td>\n",
       "      <td>102027407</td>\n",
       "      <td>102027408</td>\n",
       "      <td>rs12784396__Psoriasis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr      start       stop                          name\n",
       "0  chr11   78128047   78128048  rs881361__Alzheimer Dementia\n",
       "1   chr5  153652930  153652931     rs73281459__Schizophrenia\n",
       "2   chr5  153656738  153656739     rs73279671__Schizophrenia\n",
       "3   chr8  109260581  109260582  rs9969614__Dupuytren Disease\n",
       "4  chr10  102027407  102027408         rs12784396__Psoriasis"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_file = '../data/external/snp_list_noncancer.csv'\n",
    "snp_orig_df = pd.read_csv(snp_file, index_col=0)\n",
    "snp_orig_df['name'] = snp_orig_df.snp + '__' + snp_orig_df.disease\n",
    "diseases = snp_orig_df.disease.unique()\n",
    "snp_orig_df['stop'] = snp_orig_df['loc']+1\n",
    "snp_df = snp_orig_df[['chr', 'loc', 'stop', 'name']]\n",
    "snp_df.columns = ['chr', 'start', 'stop', 'name']\n",
    "\n",
    "\n",
    "snp_df.to_csv('../data/external/snp_list_combined.bed',sep='\\t',header=None, index=None)\n",
    "snp_bed = pybedtools.BedTool.from_dataframe(snp_df).sort()\n",
    "\n",
    "print(snp_df.shape)\n",
    "print(len(diseases))\n",
    "snp_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# melanocyte_list = ['Melanocytes','COLO_SCR_DMSO', 'COLO_SCR_PLX', 'COLO_shMITF_DMSO', 'COLO_shMITF_PLX',\n",
    "#                    'WM_SCR_DMSO', 'WM_SCR_PLX', 'WM_shMITF_DMSO', 'WM_shMITF_PLX']\n",
    "# skin_list = ['GDSD0','GDSD3','GDSD6', 'A431-CTRLi', 'A431-p63i','CAL27-CTRLi', 'CAL27-p63i',\n",
    "#              'D0-CTRLi', 'D0-p63i', 'D3-CTRLi', 'D3-p63i', 'D6-CTRLi', 'D6-p63i','SCC13-CTRLi', 'SCC13-p63i']\n",
    "# disease_to_tissue = {\n",
    "#     '':[],\n",
    "#     'paad':['Pancreas'], 'brca,ovca':['HMEC','Ovarian'], 'ovca,paad':['Ovarian','Pancreas'], \n",
    "#     'melanoma':melanocyte_list, \n",
    "#     'colon':['Colon'], \n",
    "#     'kidney,skin':['Renal']+skin_list +melanocyte_list, 'colon,lung':['Colon','Airway'], \n",
    "#     'bladder,lung,skin':['Bladder','Airway']+skin_list +melanocyte_list,\n",
    "#     'colon,endometrial':['Colon','Uterine'], 'brca,lung':['HMEC','Airway'], 'kidney':['Renal'], 'Thyroid':['Thy'], 'thyroid':['Thy'], \n",
    "#     'endometrial,ovca':['Uter','Ova'], 'skin':['GDSD0','GDSD3','GDSD6','MC'], 'brca':['HMEC'], 'esophagus':['Eso'], 'prad':['Pros'], \n",
    "#     'endometrial':['Uter'], 'ovca':['Ova'], 'lung,skin':['Air','GDSD0','GDSD3','GDSD6','MC'], 'lung,ovca':['Air','Ova'], \n",
    "#     'cervical':['Uter'], 'lung':['Air'], 'bladder':['Blad'], 'brca,melanoma':['HMEC','MC'], 'lymphoma':['GM12878'], \n",
    "#     'colon,melanoma':['Colon','MC'], 'brain':['Astro'], 'melanoma,skin':['GDSD0','GDSD3','GDSD6','MC']\n",
    "    \n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Astrocytes (3525290, 4)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H9_D0 (5982989, 4)\n",
      "H9_D10 (7871677, 4)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H9_D2 (15469189, 4)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H9_D28 (17519804, 4)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SLC_D0 (22092600, 4)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SLC_D2 (27570105, 4)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SL_D0 (30047693, 4)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mguo123/opt/anaconda3/lib/python3.7/site-packages/numpy/lib/arraysetops.py:568: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SL_D2 (35676018, 4)\n",
      "CPU times: user 8min 54s, sys: 21.7 s, total: 9min 16s\n",
      "Wall time: 6min 6s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "snp_egene_df = pd.DataFrame(columns=['TSS',\t'snp_id',\t'disease',\t'tissue'])\n",
    "for tissue in tissues:\n",
    "    # 1. annotate anchors with snp info \n",
    "    anchor = pybedtools.BedTool('../data/interim/merged/anchors_bed_sort/'+tissue+'_sort.bed')\n",
    "    anchor_snp_annon_df = anchor.intersect(snp_bed, wa=True, wb=True, sorted=True, names='anchor').to_dataframe()\n",
    "    anchor_snp_annon_df.columns = ['chr', 'start', 'stop', 'anchor', 'chr_snp','start_snp', 'stop_snp', 'snp_name']\n",
    "    anchor_snp_annon_df_filt = anchor_snp_annon_df[['anchor', 'snp_name']]\n",
    "\n",
    "    # 2. read in tss annotation of anchors\n",
    "    promoter_file = '../data/interim/annon/promoter_anchors/promoter_'+tissue+'_annon.bed'\n",
    "    if not os.path.exists(promoter_file):\n",
    "        print(promoter_file,'does not exist .. skipping')\n",
    "        continue\n",
    "    anchor_tss_annon_df = pd.read_csv(promoter_file,sep='\\t', header=None)\n",
    "    anchor_tss_annon_df.columns = ['chr_p','start_p','stop_p','TSS','chr_f','start_f','stop_f','anchor','overlap']\n",
    "    anchor_tss_annon_df_filt = anchor_tss_annon_df.sort_values('overlap', ascending=False).drop_duplicates(['TSS']).sort_index()\n",
    "    anchor_tss_annon_df_filt = anchor_tss_annon_df_filt[['anchor','TSS']]\n",
    "\n",
    "    # 3. merge with loops: read in loop information and annotate loops with snp and tss anchor\n",
    "    loop_file = '../data/interim/merged/loops/'+tissue+'.loops.csv'\n",
    "    loop_df = pd.read_csv(loop_file,index_col=0)\n",
    "    loop_df_rev = loop_df.copy()\n",
    "    loop_df_rev.columns = ['target','source','count']\n",
    "    loop_df_bi = pd.concat([loop_df, loop_df_rev],sort=False)\n",
    "    loop_df_bi = loop_df_bi.merge(anchor_snp_annon_df_filt,how='left',left_on='source',right_on='anchor' )\n",
    "    loop_df_bi.drop('anchor',axis=1,inplace=True)\n",
    "    loop_df_bi = loop_df_bi.merge(anchor_snp_annon_df_filt,how='inner',left_on='target',right_on='anchor', suffixes=['_s','_t'] )\n",
    "    loop_df_bi.drop('anchor',axis=1,inplace=True)\n",
    "    loop_df_bi = loop_df_bi.merge(anchor_tss_annon_df_filt,how='left',left_on='source',right_on='anchor' )\n",
    "    loop_df_bi.drop('anchor',axis=1,inplace=True)\n",
    "    loop_df_bi = loop_df_bi.merge(anchor_tss_annon_df_filt,how='inner',left_on='target',right_on='anchor', suffixes=['_s','_t'] )\n",
    "    loop_df_bi.drop('anchor',axis=1,inplace=True)\n",
    "\n",
    "    # 4. a SNP-egene pair where a SNP and egene are found on the same loop (regardless if found on same anchor point or different one)\n",
    "    tissue_snp_df = loop_df_bi[['snp_name_s',\t'snp_name_t',\t'TSS_s',\t'TSS_t']].fillna('')\n",
    "    tissue_snp_df['snp_name'] = tissue_snp_df.snp_name_s + '::' + tissue_snp_df.snp_name_t\n",
    "    tissue_snp_df['TSS'] = tissue_snp_df.TSS_s + '::' + tissue_snp_df.TSS_t\n",
    "    tissue_snp_df['TSS'] = tissue_snp_df.TSS.str.strip('::')\n",
    "    tissue_snp_df['snp_name'] = tissue_snp_df.snp_name.str.strip('::')\n",
    "    tissue_snp_df = tissue_snp_df[['snp_name','TSS']]\n",
    "    tissue_snp_df = pd.DataFrame(tissue_snp_df.snp_name.str.split('::').to_list(), index=tissue_snp_df.TSS).stack().reset_index('TSS')\n",
    "    tissue_snp_df.columns = ['TSS', 'snp_name']\n",
    "    tissue_snp_df = pd.DataFrame(tissue_snp_df.TSS.str.split('::').to_list(), index=tissue_snp_df.snp_name).stack().reset_index([0,'snp_name'])\n",
    "    tissue_snp_df.columns = ['snp_name', 'TSS']\n",
    "    tissue_snp_df[['snp_id','disease']]=tissue_snp_df.snp_name.str.split('__',expand=True)\n",
    "    tissue_snp_df['tissue'] = tissue\n",
    "    tissue_snp_df.drop('snp_name',axis=1,inplace=True)\n",
    "    \n",
    "    # 5. add to overall dataframe\n",
    "    snp_egene_df = pd.concat([snp_egene_df, tissue_snp_df])\n",
    "    print(tissue, snp_egene_df.shape)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_egene_df.dropna(inplace=True)\n",
    "snp_egene_df.to_csv(os.path.join(save_dir,'snp_egene_df.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_egene_normal_df = snp_egene_df[snp_egene_df.tissue.isin(normal_tissues)]\n",
    "snp_egene_normal_df.to_csv(os.path.join(save_dir, 'snp_egene_normal_df.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((33510778, 4), (33510778, 4))"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_egene_df.shape, snp_egene_normal_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. get snp egene distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "TSS_df = pd.read_csv('../data/external/TSS_annon_hg19.csv',index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_egene_df_wdist = snp_egene_df.merge(TSS_df[['chr','loc','TSS']], how='left', on='TSS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5min 35s, sys: 3min 3s, total: 8min 38s\n",
      "Wall time: 9min 10s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# takes awhile (~1-2 min)\n",
    "snp_egene_df_wdist['name'] = snp_egene_df_wdist.snp_id + '__' + snp_egene_df_wdist.disease\n",
    "snp_egene_df_wdist = snp_egene_df_wdist.merge(snp_df[['chr','start','name']], how='left', on='name')\n",
    "snp_egene_df_wdist.dropna(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(355476582, 9)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_egene_df_wdist.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(355474873, 9)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_egene_df_wdist = snp_egene_df_wdist[snp_egene_df_wdist.chr_x==snp_egene_df_wdist.chr_y]\n",
    "snp_egene_df_wdist.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_egene_df_wdist['distance'] = abs(snp_egene_df_wdist['loc'] - snp_egene_df_wdist['start'])\n",
    "snp_egene_df_wdist = snp_egene_df_wdist[['TSS',\t'snp_id',\t'disease',\t'tissue', 'distance']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    3.554749e+08\n",
       "mean     1.446384e+05\n",
       "std      3.312608e+05\n",
       "min      1.000000e+00\n",
       "25%      9.490000e+03\n",
       "50%      1.202700e+04\n",
       "75%      7.590800e+04\n",
       "max      2.024192e+06\n",
       "Name: distance, dtype: float64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_egene_df_wdist.distance.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fc87572d0d0>"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZwAAAEGCAYAAABRvCMcAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3de3xdZZ3v8c8veydpm17SJiktvdBCg5g6WCC0isLxgEoZL8VjHVs4iiNHvMA5KmdmoC9Hx2GYGTtzztTjDCgoiPKilsrI0Jc3FEEcL7RNoRRaKE0vtOk1vaRJm+ve+Z0/1hPY3d072U13Vqp+36/XfmXtZz3P73n2ytr7l7XWk7XN3RERERlqJcM9ABER+eOghCMiIrFQwhERkVgo4YiISCyUcEREJBbJ4R7Amaa6utpnzJgx3MMQEfm9sm7duoPuXtNfHSWcLDNmzKChoWG4hyEi8nvFzF4dqI5OqYmISCyUcEREJBZKOCIiEgslHBERiYUSjoiIxEIJR0REYlFQwjGz+Wa22cwazez2HOvLzezhsH61mc3IWLcklG82s6sHimlmM0OMLSFmWSi/wsyeNbOUmS3MqP9fzWx9xqPTzK4N6x4ws+0Z6+YMZiOJiMjpGzDhmFkCuAu4BqgDFptZXVa1G4Ej7j4LWAYsDW3rgEXAbGA+cLeZJQaIuRRY5u61wJEQG2An8DFgeWbH7v6Uu89x9znAlUA78LOMKn/Zt97d1w/0ekVEZGgUcoQzF2h0923u3g2sABZk1VkAfCcsPwJcZWYWyle4e5e7bwcaQ7ycMUObK0MMQsxrAdx9h7tvAHr7GetC4Cfu3l7A6xIRkRgVcqeBKcCujOdNwLx8ddw9ZWZHgapQ/kxW2ylhOVfMKqDF3VM56hdiEfAvWWV/b2ZfAn4B3O7uXacQ75QtX70zZ/l186YPZbciIme8Qo5wLEdZ9teE5qtTrPIBmdlk4E+AxzOKlwAXAJcCE4Db8rS9ycwazKyhubm5kO5EROQUFZJwmoBpGc+nAnvy1TGzJDAOONxP23zlB4HKECNfX/n8GfCou/f0Fbj7Xo90Ad8mOpV3Ene/193r3b2+pqbfe8+JiMggFZJw1gK1YfZYGdFpq1VZdVYBN4TlhcCT7u6hfFGYxTYTqAXW5IsZ2jwVYhBiPlbga1kMfC+zIBz1EK4NXQu8WGAsEREpsgGv4YRrMrcQnapKAPe7+0YzuwNocPdVwH3Ag2bWSHRksyi03WhmK4FNQAq42d3TALlihi5vA1aY2Z3AcyE2ZnYp8CgwHnifmf2tu88O62YQHTE9nTX8h8yshuhU3XrgU6e4fUREpEgsOqiQPvX19X46X0+gSQMi8sfIzNa5e31/dXSnARERiYUSjoiIxEIJR0REYqGEIyIisVDCERGRWCjhiIhILJRwREQkFko4IiISCyUcERGJhRKOiIjEQglHRERioYQjIiKxUMIREZFYKOGIiEgslHBERCQWSjgiIhILJRwREYmFEo6IiMRCCUdERGKhhCMiIrFQwhERkVgUlHDMbL6ZbTazRjO7Pcf6cjN7OKxfbWYzMtYtCeWbzezqgWKa2cwQY0uIWRbKrzCzZ80sZWYLs/pPm9n68Fg1UCwREYnfgAnHzBLAXcA1QB2w2MzqsqrdCBxx91nAMmBpaFsHLAJmA/OBu80sMUDMpcAyd68FjoTYADuBjwHLcwyzw93nhMf7M8rzxRIRkZgVcoQzF2h0923u3g2sABZk1VkAfCcsPwJcZWYWyle4e5e7bwcaQ7ycMUObK0MMQsxrAdx9h7tvAHoLeWH9xRIRkfgVknCmALsynjeFspx13D0FHAWq+mmbr7wKaAkx8vWVywgzazCzZ8ysL6kUHMvMbgrtG5qbmwvoTkRETlWygDqWo8wLrJOvPFei66/+QKa7+x4zOxd40sxeAFoLjeXu9wL3AtTX1xfSn4iInKJCjnCagGkZz6cCe/LVMbMkMA443E/bfOUHgcoQI19fJ3H3PeHnNuCXwEWDjSUiIkOjkISzFqgNM77KiCYBrMqqswq4ISwvBJ50dw/li8IstplALbAmX8zQ5qkQgxDzsf4GZ2bjzaw8LFcDbwM2DSaWiIgMnQETTrgGcgvwOPASsNLdN5rZHWbWNyPsPqDKzBqBW4HbQ9uNwEpgE/BT4GZ3T+eLGWLdBtwaYlWF2JjZpWbWBHwIuMfM+uq/EWgws+eJEsxX3H1Tf7FERCR+Fh0ISJ/6+npvaGgYdPvlq3fmLL9u3vRBxxQROdOZ2Tp3r++vju40ICIisVDCERGRWCjhiIhILJRwREQkFko4IiISCyUcERGJhRKOiIjEQglHRERioYQjIiKxUMIREZFYKOGIiEgslHBERCQWSjgiIhILJRwREYmFEo6IiMRCCUdERGKhhCMiIrFQwhERkVgo4YiISCyUcEREJBYFJRwzm29mm82s0cxuz7G+3MweDutXm9mMjHVLQvlmM7t6oJhmNjPE2BJiloXyK8zsWTNLmdnCjPpzzOx3ZrbRzDaY2Ycz1j1gZtvNbH14zDnVDSQiIsUxYMIxswRwF3ANUAcsNrO6rGo3AkfcfRawDFga2tYBi4DZwHzgbjNLDBBzKbDM3WuBIyE2wE7gY8DyrL7bgY+6e18fXzWzyoz1f+nuc8Jj/UCvV0REhkYhRzhzgUZ33+bu3cAKYEFWnQXAd8LyI8BVZmahfIW7d7n7dqAxxMsZM7S5MsQgxLwWwN13uPsGoDezY3d/xd23hOU9wAGgpuAtICIisSgk4UwBdmU8bwplOeu4ewo4ClT10zZfeRXQEmLk6ysvM5sLlAFbM4r/PpxqW2Zm5YXGEhGR4iok4ViOMi+wTrHKB2Rmk4EHgT93976joCXABcClwATgtjxtbzKzBjNraG5uLqQ7ERE5RYUknCZgWsbzqcCefHXMLAmMAw730zZf+UGgMsTI19dJzGws8CPgr939mb5yd9/rkS7g20Sn8k7i7ve6e72719fU6GyciMhQKCThrAVqw+yxMqJJAKuy6qwCbgjLC4En3d1D+aIwi20mUAusyRcztHkqxCDEfKy/wYX2jwLfdffvZ62bHH4a0bWgFwt4vSIiMgQGTDjhesotwOPAS8BKd99oZneY2ftDtfuAKjNrBG4Fbg9tNwIrgU3AT4Gb3T2dL2aIdRtwa4hVFWJjZpeaWRPwIeAeM+ur/2fAFcDHckx/fsjMXgBeAKqBOwexjUREpAgsOqiQPvX19d7Q0DDo9stX78xZft286YOOKSJypjOzde5e318d3WlARERioYQjIiKxUMIREZFYKOGIiEgslHBERCQWSjgiIhILJRwREYmFEo6IiMRCCUdERGKhhCMiIrFQwhERkVgo4YiISCyUcEREJBZKOCIiEgslHBERiYUSjoiIxEIJR0REYqGEIyIisVDCERGRWCjhiIhILJRwREQkFgUlHDObb2abzazRzG7Psb7czB4O61eb2YyMdUtC+WYzu3qgmGY2M8TYEmKWhfIrzOxZM0uZ2cKs/m8I9beY2Q0Z5ZeY2Quhj6+ZmZ3KxhERkeIZMOGYWQK4C7gGqAMWm1ldVrUbgSPuPgtYBiwNbeuARcBsYD5wt5klBoi5FFjm7rXAkRAbYCfwMWB51vgmAH8DzAPmAn9jZuPD6q8DNwG14TF/oNcrIiJDo5AjnLlAo7tvc/duYAWwIKvOAuA7YfkR4KpwNLEAWOHuXe6+HWgM8XLGDG2uDDEIMa8FcPcd7r4B6M3q+2rg5+5+2N2PAD8H5pvZZGCsu//O3R34bl8sERGJXyEJZwqwK+N5UyjLWcfdU8BRoKqftvnKq4CWECNfX4WOb0pY7m/cAJjZTWbWYGYNzc3NA3QnIiKDUUjCyXXdwwusU6zy/px2LHe/193r3b2+pqZmgO5ERGQwCkk4TcC0jOdTgT356phZEhgHHO6nbb7yg0BliJGvr0LH1xSW+xu3iIjEpJCEsxaoDbPHyogmAazKqrMK6JsdthB4Mlw3WQUsCrPYZhJduF+TL2Zo81SIQYj52ADjexx4t5mND5MF3g087u57gTYze0u4NvTRAmKJiMgQGTDhhOsptxB9sL8ErHT3jWZ2h5m9P1S7D6gys0bgVuD20HYjsBLYBPwUuNnd0/lihli3AbeGWFUhNmZ2qZk1AR8C7jGzjaGPw8DfESWxtcAdoQzg08C3iCYrbAV+MohtJCIiRWDRQYX0qa+v94aGhkG3X756Z87y6+ZNH3RMEZEznZmtc/f6/uroTgMiIhILJRwREYmFEo6IiMRCCUdERGKhhCMiIrFQwhERkVgo4YiISCyUcEREJBZKOCIiEgslHBERiYUSjoiIxEIJR0REYqGEIyIisVDCERGRWCjhiIhILJRwREQkFko4IiISCyUcERGJhRKOiIjEQglHRERiUVDCMbP5ZrbZzBrN7PYc68vN7OGwfrWZzchYtySUbzazqweKaWYzQ4wtIWZZf32Y2fVmtj7j0Wtmc8K6X4Y++tZNHNxmEhGR0zVgwjGzBHAXcA1QByw2s7qsajcCR9x9FrAMWBra1gGLgNnAfOBuM0sMEHMpsMzda4EjIXbePtz9IXef4+5zgI8AO9x9fcbYru9b7+4HCt4yIiJSVIUc4cwFGt19m7t3AyuABVl1FgDfCcuPAFeZmYXyFe7e5e7bgcYQL2fM0ObKEIMQ89oB+si0GPheAa9JRERiVkjCmQLsynjeFMpy1nH3FHAUqOqnbb7yKqAlxMjuK18fmT7MyQnn2+F02hdzJCgAzOwmM2sws4bm5uZcVURE5DQVknByfUh7gXWKVT7gOMxsHtDu7i9mrL/e3f8EuDw8PpIjBu5+r7vXu3t9TU1NrioiInKaCkk4TcC0jOdTgT356phZEhgHHO6nbb7yg0BliJHdV74++iwi6+jG3XeHn23AcqJTeSIiMgwKSThrgdowe6yM6IN9VVadVcANYXkh8KS7eyhfFGaYzQRqgTX5YoY2T4UYhJiPDdAHZlYCfIjoWhChLGlm1WG5FHgvkHn0IyIiMUoOVMHdU2Z2C/A4kADud/eNZnYH0ODuq4D7gAfNrJHoqGNRaLvRzFYCm4AUcLO7pwFyxQxd3gasMLM7gedCbPL1EVwBNLn7toyycuDxkGwSwBPAN09h24iISBFZOEiQoL6+3hsaGgbdfvnqnTnLr5s3fdAxRUTOdGa2zt3r+6ujOw2IiEgslHBERCQWSjgiIhILJRwREYmFEo6IiMRCCUdERGKhhCMiIrFQwhERkVgo4YiISCyUcEREJBZKOCIiEgslHBERiYUSjoiIxEIJR0REYqGEIyIisVDCERGRWCjhiIhILJRwREQkFko4IiISCyUcERGJhRKOiIjEoqCEY2bzzWyzmTWa2e051peb2cNh/Wozm5Gxbkko32xmVw8U08xmhhhbQsyy/vowsxlm1mFm68PjGxmxLjGzF0Kbr5mZnfomEhGRYhgw4ZhZArgLuAaoAxabWV1WtRuBI+4+C1gGLA1t64BFwGxgPnC3mSUGiLkUWObutcCREDtvH8FWd58THp/KKP86cBNQGx7zB3q9IiIyNAo5wpkLNLr7NnfvBlYAC7LqLAC+E5YfAa4KRxMLgBXu3uXu24HGEC9nzNDmyhCDEPPaAfrIycwmA2Pd/Xfu7sB3M2KJiEjMCkk4U4BdGc+bQlnOOu6eAo4CVf20zVdeBbSEGNl95esDYKaZPWdmT5vZ5Rn1mwYYNwBmdpOZNZhZQ3Nzc64qIiJymgpJOLmOIrzAOsUq76+PvcB0d78IuBVYbmZjCxx3VOh+r7vXu3t9TU1NrioiInKaCkk4TcC0jOdTgT356phZEhgHHO6nbb7yg0BliJHdV84+wum6QwDuvg7YCpwf6k8dYNwiIhKTQhLOWqA2zB4rI5oEsCqrzirghrC8EHgyXDdZBSwKM8xmEl24X5MvZmjzVIhBiPlYf32YWU2YhICZnRv62Obue4E2M3tLuNbz0YxYIiISs+RAFdw9ZWa3AI8DCeB+d99oZncADe6+CrgPeNDMGomObBaFthvNbCWwCUgBN7t7GiBXzNDlbcAKM7sTeC7EJl8fwBXAHWaWAtLAp9z9cFj3aeABYCTwk/AQEZFhYNFBhfSpr6/3hoaGQbdfvnpnzvLr5k0fdEwRkTOdma1z9/r+6uhOAyIiEgslHBERiYUSjoiIxEIJR0REYqGEIyIisVDCERGRWCjhiIhILJRwREQkFko4IiISCyUcERGJhRKOiIjEQglHRERioYQjIiKxUMIREZFYKOGIiEgslHBERCQWSjgiIhILJRwREYmFEo6IiMRCCUdERGJRUMIxs/lmttnMGs3s9hzry83s4bB+tZnNyFi3JJRvNrOrB4ppZjNDjC0hZll/fZjZu8xsnZm9EH5emRHrl6GP9eEx8dQ3kYiIFMOACcfMEsBdwDVAHbDYzOqyqt0IHHH3WcAyYGloWwcsAmYD84G7zSwxQMylwDJ3rwWOhNh5+wAOAu9z9z8BbgAezBrb9e4+JzwODLhFRERkSBRyhDMXaHT3be7eDawAFmTVWQB8Jyw/AlxlZhbKV7h7l7tvBxpDvJwxQ5srQwxCzGv768Pdn3P3PaF8IzDCzMoL3QAiIhKPQhLOFGBXxvOmUJazjrungKNAVT9t85VXAS0hRnZf+frI9EHgOXfvyij7djid9sWQ0E5iZjeZWYOZNTQ3N+eqIiIip6mQhJPrQ9oLrFOs8gHHYWaziU6zfTJj/fXhVNvl4fGRHDFw93vdvd7d62tqanJVERGR01RIwmkCpmU8nwrsyVfHzJLAOOBwP23zlR8EKkOM7L7y9YGZTQUeBT7q7lv7grr77vCzDVhOdCpPRESGQSEJZy1QG2aPlRFNAliVVWcV0QV7gIXAk+7uoXxRmGE2E6gF1uSLGdo8FWIQYj7WXx9mVgn8CFji7r/pG5CZJc2sOiyXAu8FXizg9YqIyBBIDlTB3VNmdgvwOJAA7nf3jWZ2B9Dg7quA+4AHzayR6KhjUWi70cxWApuAFHCzu6cBcsUMXd4GrDCzO4HnQmzy9QHcAswCvmhmXwxl7waOA4+HZJMAngC+ecpbSEREisKigwrpU19f7w0NDYNuv3z1zpzl182bPuiYIiJnOjNb5+71/dXRnQZERCQWSjgiIhILJRwREYmFEo6IiMRCCUdERGKhhCMiIrFQwhERkVgo4YiISCwGvNOADN6elg5e3tfGrsPt/OKl/dxx7ZuYUjlyuIclIjIsdIQzRNyd+3+znSde2s/h9m7WbD/Mgn/7NetePTzcQxMRGRZKOEOktTNFe3ea9144mc+/83wevfkyRpcnWXzvan695eBwD09EJHZKOEPkQGsnAJPGjgBgzfYj/Pe3nMO4UaXcvPxZ7v/19rz3XRMR+UOkhDNEDrRFXzo6MSQcgFFlSRZePJXWjh5+/MLe4RqaiMiw0KSBIbK/tZNRZQlGl5+4iadNGMUV59fw9CvNjBlRygu7Wzja0cOSa97ItAmjhmm0IiJDTwlniBxo6+KsjKObTFddMJGX9rby1OYDjBtZSirdy4u7W1n5ybcyaVzuNiIiv+90Sm0IuDsH2jqZOKY85/pkooRPXnEe//td57P+S+/ioU+8hcPHu7n+W8+w72hnUcbw9CvN/HzT/qLEEhEpBiWcIdDamaKzp/eE6zfZRpYlqBpdzvfW7GLTnlYWz53OzsPtvO0rT/K+f/01v916kHxfjufuHDnezdbmY/Ske09av6elg089uI5PfLeBv//RJlI56oiIxE2n1IZA3wy1s/Ic4eQys7qC/3llLau3HeLZnS1c983VnFtTwXVzp7No7nRGlydxd+751Ta++sQrdPZESeSNk8fy9esvZkZ1xWux/uHHL9HrzsJLpvLN/9zO77YdIlFSwu4j7XzwkqncPv8CzKy4LxrYd7ST6tFlJBP6O0ZETqaEMwRyzVArRPXoct5z4dm8e/YkRpcnWb5mJ3f+6CXu/uVWPvOO81j36hF+8uI+Lpg0hvNqRmMGv3jpAFd/9Vcs/eCFvOfCyazdcZgfbtjL5995Pp99Zy29vc6vGw8ydkQpEyrKuOfpbWzYdZTv3jiXHzzbxMqGJv7q6jcw79yqUxpr39FXX+J6ZtshPnLfat75xrO4+/qLhyShicjvN8t32uaPVX19vTc0NAy6/fLVO/nBs01s2tvKX7+n7rTH03SknZ9t2k/jgWMYMP9Nk3j7rOrXPtCPtHfzvTU7aTrSwaiyBIkSI1lifO6d51OadaTh7vzkxX38uvEgEyrKOHy8m/JkCeXJEn7wmcuYNXEM7s7BY91UVZRRUpI7aWxtPsbNDz1LWbKEf1p4IeXJBB+4+zeke522zhR/+/7Z3HDZDJ5+pZmv/WILX3pvHW+eVnna20JEzlxmts7d6/uto4RzomIknG88vZVEifGJy88t2ri2HzxOssRyTp1O9zqv7G/j+aYWth44xsJLpvKGSWNzxnF3nnhpP1sOHOMd59cwadxIvv70VsoSxhXn17B5Xxsv72tjVFmCWRNHUztxDOefNZpzqiqYUFHGnpYO/vo/XqQsWYIBLR09VJQlSPU6n/4v5/HDDXtpbD7G4rnTeGj1Ttxh6viR/PizlzN2RGnRtscfk31HO0n19jJ1vKbNy5mraAnHzOYD/w9IAN9y969krS8HvgtcAhwCPuzuO8K6JcCNQBr4X+7+eH8xzWwmsAKYADwLfMTdu4vZR38Gm3BWPb+HnlQvnT1p/u5Hm3jz1EoWzJlyynGGw67D7Xzr19voSTuTxo7gzdMqaevs4UBrF/vbOmnrTJ1Qf+r4kVw3dzrJRAmr1u9m8/42Pv62mZxTVUF7V4p/faqRox09XDStkoumj+eB325n9tnj+PCl05g+YRS/3XqIsmQJFWUJJo4t5+xxI+lO99Kw4whbDrRx2XnVvO/Cs9l28Bj/8vNXeG5nCwsvmconrjiX/a2d/PD5vexv7WRG9ShqJ47hsllVTBwTnb48fLybXYfbmVFdwbiRpaR7/bWbqG5oamHHoXYuO6+Ka940icpRZcOxufu14+BxGg8c4/LzqylPJnhi034+//B6HPjWDfW85RRPfcbJ3flN4yE27G5h/KgyqirKmFFdwTlVoyhPJoZ7eH/QOnvSjCgd3m1clIRjZgngFeBdQBOwFljs7psy6nwGuNDdP2Vmi4APuPuHzawO+B4wFzgbeAI4PzTLGdPMVgI/cPcVZvYN4Hl3/3ox++jv9Q424Xz8gbU8+fIBZlZXsP3gcd735rN56xn84ZCt6Ug73aleZlZXnHT9paM7zeH2btq7U6TSTu3E0SdMDEj19pIsef15c1sXzW1dvHHyGMyMX24+wM827WdMeZK2rhQlBr05drsSg4ljRrCvtZPShNGTdkaVJZhZXcFLe1tfa5MoMSpHltLS0UM6FL556jh60s6mva2vxaseXUZrR4ruMEuvxKCiLBpDwoxxo6IjrspRpVSPLqdmdDmjy5OMLEvQ2ZOm6UgHB491Mbo8SeWoMrrTvRw61kV7d5qK8uifejt7emnt7KGiLMnss8dybk0FB1q72HWkndHlpVwwaQyTxo2gpb2bg8e6OXS8i0PHuqPH8S46e3p526wqrrzgLJ5+pZmVDbtI9zoV5UnOq6lgQ9NRplRGCbm1oyecwixhQ9NRDh3rpr0nTUd3iuNd6deW27vTjC5PcsGkMUybMIr9rZ3sbukgUVJCdUUZE8eOYGb1KKaOH0VzWxc7Dh3neFeKEaUJzIzmtk72He0kmShhwqgyKsqT9LrTleplx8HjbDlwjDEjkrzjDTXMnTGB7nQvB4918/2GXby8ry3n7/XcmtFcPL2Sc6oq2LK/jc37jzGhopS6yWMpTZSwZvthNu1tZWZ1BRdPH8/M6grGjEgyqixJqrf3tX2hclQppYkSjrb3cLw79dq1ydJECT3pXnrSvXSne+nq6WX7weO8vK+VnrRz8fRKLpxaSa87x7pSlCZKqB5dhpnx4u6jbNrTSknYrwD2t3Zx+HgXI8M/b5sZ3aleulJpulK9pNLO9AmjeNOUsVSPLufQ8W6OtveQKDHKkiWUJqJT1eWlJYwbWcrYEaV09qRp6ejBHSZUlDKyLMnuIx28eug4DowfVcaosO+1d6fp6EnT2ZNmZGmC8yaOZuKYcl7e28aG3Ufp7XXGV5RxvCvFL14+wPO7Wnjj5LEsvGQqb5w8hn1HO2nt6OENk8Zy4dRx7G/t5Jlthzl4rIv6c8Zz4bRKNu1p5VevNOM4V9TWcPE54086DX8qipVw3gp82d2vDs+XALj7P2bUeTzU+Z2ZJYF9QA1we2bdvnqh2Ukxga8AzcAkd09l9l2sPjLHnctgE0661/nu73bwDz9+iZ608z/ePpNza0afcpw/RL3uPLx2F8e7U1w6YwKzJ4+lpMToSfXS2pmipaMbw5g6fiTlyRL2HO3k+V0tjC5PMm/mBMpLE7S0d7Pu1SOMryijbvJYRpQmSPX2cqC1i83729i8r41kib32xjx0rJvmY12MKktQPbqciWPKmTxuJKUJY8/RTl5oaqE1HLmlep3jXSnaOlN0p9J0p3tJlJQwflQpY0aU0pVK096VJpkwRpcnKUuW0NXTS2cqTVmihPLSBJ3daXa3dNDRkyZZYoyvKKOzO01b1+tHh0Y0Hb6iPMno8iQV5Ulwp7H5GJ09vSTMmDtzArMmjmbtjsNs3tfGRdOjI+XuVC8P/HYHu1s6gOhDvG8sZYkSSsPPvuft3Wn2tXZytKOHMSOSVI4sxYFjXSnaOlKkM973BiRDggeoKE8ydkQSdzjenaI71YsZJMyYUFHGxDEjONaVYmvzMVIZfzmcNbact8+qZvbZ4+jsSXOsK8XBY900t3Wxp6WDnYfb6ehJM3ZEkknjRnC8K83+1k563ZlSOZKzK0dy8FgXu4500J0qzlT+irIEJWYn/B5ySZQY7v7aHzVliRIqyqNTxZ09aQCSJSUkS4xEwigxo6W9O+cfTkOt78/Bvq6njh/52h+6TUc6TilWosQwovfAmPIkD3/yrdSdnft0/IDjKiDhFDJLbQqwK+N5EzAvX52QKI4CVaH8mTg54WUAAAmlSURBVKy2feeZcsWsAlrcPZWjfrH6OImZ3QTcFJ4eM7PNueoVqBo4+MWlpxGh+KqBM+IW1U+9vljQmO4ZfPzBGtZttS3r+Q7g0TPo95flhHG9CqwpsOELWc+3F2tEvyfbqpheBX4zuKYnjWn2Hac1lHMGqlBIwsk1VSk7r+erk68813Fbf/WL2cfJhe73AvfmWneqzKxhoCwfN42pcGfiuM7EMcGZOa4zcUxwZo5rOMZUyAm7JmBaxvOpwJ58dcLprnHA4X7a5is/CFSGGNl9FasPEREZBoUknLVArZnNNLMyYBGwKqvOKuCGsLwQeNKji0OrgEVmVh5mn9USHXXnjBnaPBViEGI+Vsw+CtssIiJSbAOeUgvXS24BHieaXny/u280szuABndfBdwHPGhmjURHHYtC241h1tkmIAXc7O5pgFwxQ5e3ASvM7E7guRCbIvcxlIpyaq7INKbCnYnjOhPHBGfmuM7EMcGZOa7Yx6R//BQRkVjoLosiIhILJRwREYmHu+tRhAcwH9gMNAK3FynmNKJJFC8BG4HPhvIvA7uB9eHxpxltloQxbAauHmh8wExgNbAFeBgoC+Xl4XljWD8jo80Oon+lWE90HQ+iWxH9PMT5OTA+lBvwtRBnA3BxRpwbQv0twA0Z5ZeE+I2hrfXXR1j3hoztsR5oBT43TNtqJXAAeDGj7bBtn9DHRqJrnJ19fQD/DLwc+n0UqAzlM4COjG32jdPsO9/r+1UYU1ff6xum39eMrD5aw7h2ZJQ/nDGmHcD6mLfV54F2oJtoBu5nTyNWMfernH3k/Uwbjg/nP7QH0aSErcC5QBnwPFBXhLiTef3DYQzRrXrqwpvyL3LUrwt9l4c329YwtrzjI/pwXBSWvwF8Oix/pu/NQzRB4+GMfnYA1Vl9/xPhzU5094elYflPgZ+EnfMtwOqMnXhb+Dk+LPftyGuAt4Y2PwGu6a+PPL+PfUT/iDYc2+pJ4GJOTDjDtn1CH8+EMW3N6OPdQDIsL82oPyNz7FnbbTB953t9e4B3EE342RZe53D8vh7O6uMq4L1EiTCRYyz/F/hSzNvq1bCtxhP9n2xjGO9w71cn9dHvZ1qxP3z/GB/hl/R4xvMlwJIh6OcxonvD5XtTntAv0Qy9t+YbX9hRDvL6h85r9frahuVkqNf3F88OTk44m4HJYXkysDks30N0D7sT6gGLgXsyyu8JZZOBlzPKX6uXr48c2+HdwG/C8nBtqxmcmHCGbftktJ0BvJhZLyPOB4CHwvIJY8+oM+i+872+jDH1jXG4fl+W2UcYV1tfvYz4RnQHk9q4t1XWvtBA9Fkw7PtVrn0830PXcIoj1+1/inqraDObAVxEdAoA4BYz22Bm95vZ+AHGka+84FsJAX23EoLojg0/M7N14bZAAGe5+95Qfy8wcZBjmhKWs8v76yPbIqIbuvYZjm01nhMN5/YpZP/8ONFfq31mmtlzZva0mV2eEadYffc3puHat7Nj9XDydroc2O/uWzLK4t5Wx4DziD4LzvT96gRKOMVRyO1/Bh/cbDTw78Dn3L0V+DrRDjcH2Et0iN/fOE61vL9YAG9z94uBa4CbzeyK/oZfxDENKPyT7/uB74ei4d5WAw65iOM41T6ilWZfILpm8VAo2gtMd/eLgFuB5WY2tsh95ysfzt9XIa9vMSf+MRPrtgqfBYuAx8JnwaBjFVjen1Nuo4RTHEN2Gx0zKyVKNg+5+w8A3H2/u6fdvRf4JtFXM/Q3jmLeSgh33xN+HiC62DwX2G9mk0P9yUQXzQczpqawnF1OP31kugZ41t33D/O2aska13Bun7z7p5ndQHS94noP50XcvcvdD4XldUTXSM4vct85y4d5386OVZrRpq/ufyOaQEDc2yrjs+BVotOCg46Vp7xo+1Ve/Z1v06OwB9F54G1EFzP7LlzOLkJcI/rSua9mlU/OWP48sCIsz+bEC6vbiC6q5h0f0ZFA5oXVz4TlmznxwurKsFwBjMlY/i3RLKF/5sQLi/8Ult/DiRcW14TyCUQXP8fz+oXQCWHd2lC37+Lln4bynH1kbZsVwJ8P97bi5Gs4w7Z9MvqYQfSB2NfHfKIL9jVZ27CGcLGc6GL87iL0ne/1XRjGsD2UDee+ndnH24lmhCUyxjMfeHoYt9UKoiPAzH3hTNivTuij38+0OD+Y/5AfRDM2XiF6Q3+hSDHfTnSIuoGMaaLAg0TTFzcQ3R8u8036hTCGzYRZJv2NL7xJ1hDNevk+UB7KR4TnjWH9uRn1nw+PjX2xiM6B/4Jo6uQvMnZiA+4K/b4A1Gf0/fEQv5ETk0Q90YXkrcC/8fpkhZx9ZLQbRfRtsOMyyoZjW60iOtXSQ/RX4I3DuX1CH68QnTZzYH8YUyPROfgTpvQCHwy/2+eJvnX3fafZd77XtzpjTIfDmIZt387ooy2MK9X3+wvrHgA+lbXPxbWt/jFspy5gJ69/Fgz3fpWzj3wP3dpGRERioWs4IiISCyUcERGJhRKOiIjEQglHRERioYQjIiKxGPAbP0Xk9JnZl4luSTIW+JW7P5Gn3rXAK+6+KcbhicRCRzgiMXL3L+VLNsG1RHcBFvmDo4QjMkTM7AtmttnMniD6vh7M7AEzWxiWv2Jmm8KNKv+PmV1GdB+4fzaz9WZ2npl9wszWmtnzZvbvZjYqI87XzOy3ZratL2ZY91dm9kJo85VQdp6Z/TTccPU/zeyC2DeI/NHTKTWRIWBmlxDdNuUiovfZs8C6jPUTiL4S4AJ3dzOrdPcWM1sF/NDdHwn1Wtz9m2H5TqL/xv/XEGYy0d0oLiD6r/xHzOwaoqOkee7eHvoBuJfov+S3mNk84G7gyiHcBCInUcIRGRqXA4+6eztASCSZWom+ffNbZvYj4Id54rwpJJpKYDSv37QR4D88usnlJjM7K5S9E/h2X7/ufjjcYfgy4Ptmr93gt/y0Xp3IICjhiAydvPeNcveUmc0l+nbJRcAt5D7ieAC41t2fN7OPEX3rY5+ujGXL+JndbwnRd8PMOZXBixSbruGIDI1fAR8ws5FmNgZ4X+bKcNQxzt1/DHyO6PtfILpx5JiMqmOAveHW9NcX0O/PgI9nXOuZ4NH3pmw3sw+FMjOzN5/GaxMZFCUckSHg7s8SfW/KeqLvMPnPrCpjgB+a2QbgaaJb8UN0C/q/DN8geR7wRaK7Kv8ceLmAfn9KdD2nwczWA38RVl0P3GhmfXf5XnAaL09kUHS3aBERiYWOcEREJBZKOCIiEgslHBERiYUSjoiIxEIJR0REYqGEIyIisVDCERGRWPx/ZajMvOB0ALYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.distplot(snp_egene_df_wdist.distance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_egene_df_wdist.to_csv(os.path.join(save_dir,'snp_egene_df_wdist.csv'))\n",
    "snp_egene_normal_df_wdist = snp_egene_df_wdist[snp_egene_df_wdist.tissue.isin(normal_tissues)]\n",
    "snp_egene_normal_df_wdist.to_csv(os.path.join(save_dir, 'snp_egene_normal_df_wdist.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
